{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "589_I_6aOC5q",
        "w9m-oNNxiJkm"
      ]
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "589_I_6aOC5q"
      },
      "source": [
        "#Química das/nas Nuvens\n",
        "\n",
        "A química da atmosfera oferece novos pontos de vista sobre questões importantes nos dias de hoje, tais como o **aquecimento global**. Este _notebook_ oferece uma forma lúdica e interativa de compreender os fenómenos por detrás do efeito de estufa e de como diferentes tipos de emissões têm um impacto diferenciado no aquecimento global. No final do dia, são apenas moléculas a fazer o seu dia-a-dia, vibrando alegremente na atmosfera ao sabor dos fotões que recebem do Sol. Cabe a nós reflectir sobre o seu impacto nas nossas vidas!"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Instruções Básicas\n"
      ],
      "metadata": {
        "id": "4TS77Tu_Z8O9"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Preparação do _notebook_"
      ],
      "metadata": {
        "id": "VTd__zURaFo0"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lrLPTNtfv1jA"
      },
      "source": [
        "%%capture\n",
        "%%bash\n",
        "# Install RDKit and xTB. Takes 2-3 minutes\n",
        "#wget -c https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
        "#chmod +x Miniconda3-latest-Linux-x86_64.sh\n",
        "#bash ./Miniconda3-latest-Linux-x86_64.sh -b -f -p /usr/local\n",
        "#conda install  -y -c conda-forge python\n",
        "#conda install  -y -c conda-forge rdkit\n",
        "#conda install  -y -c conda-forge xtb\n",
        "pip install py3Dmol\n",
        "pip install seaborn\n",
        "\n",
        "# install XTB\n",
        "wget -c https://github.com/grimme-lab/xtb/releases/download/v6.5.1/xtb-6.5.1-linux-x86_64.tar.xz\n",
        "tar xf xtb-6.5.1-linux-x86_64.tar.xz\n",
        "export PATH=/content/xtb-6.5.1/bin:${PATH}\n",
        "cd ..\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7kssawSHxzAq"
      },
      "source": [
        "import sys\n",
        "sys.path.append('/usr/local/lib/python3.7/site-packages/')\n",
        "import os\n",
        "\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import time\n",
        "#from rdkit import Chem\n",
        "#from rdkit.Chem import AllChem\n",
        "from collections import defaultdict\n",
        "#from rdkit.Chem import rdFMCS\n",
        "#from rdkit.Chem import Draw\n",
        "#from rdkit.Chem.Draw import IPythonConsole\n",
        "#from rdkit.Chem import rdDistGeom\n",
        "#IPythonConsole.ipython_3d = True\n",
        "\n",
        "import py3Dmol\n",
        "import matplotlib.pyplot as plt\n",
        "import subprocess\n",
        "import seaborn as sb\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xZ0woKD8x5Pp"
      },
      "source": [
        "class Molecule():\n",
        "  def __init__(self,fn=None, name=None):\n",
        "    self.name='None'\n",
        "    self.symbols=list()\n",
        "    self.geo=np.array([])\n",
        "    if fn:\n",
        "      self.read_xyz(fn)\n",
        "    if name:\n",
        "      self.name=name\n",
        "\n",
        "  def natoms(self):\n",
        "      return len(self.symbols)\n",
        "  \n",
        "  def show(self): \n",
        "    p = py3Dmol.view(width=600,height=500)\n",
        "    p.addModel(self.write_xyz())\n",
        "    #p.setStyle({'stick':{}})\n",
        "    p.setStyle({'stick':{},'sphere':{'scale':0.3}})\n",
        "    p.setBackgroundColor('0xeeeeee')\n",
        "    p.zoomTo()\n",
        "    p.show()\n",
        "\n",
        "  def read_xyz(self, fn):\n",
        "    \"Reads a XYZ file\"\n",
        "    with open(fn,'r') as f:\n",
        "      data = f.readlines()\n",
        "    natoms = int(data[0])\n",
        "    g = list()\n",
        "    for i in range(2,2+natoms):\n",
        "      l = data[i].split()\n",
        "      self.symbols.append(l[0].capitalize())\n",
        "      g.append(list(map(float,l[1:4])))\n",
        "    self.geo=np.array(g)\n",
        "\n",
        "  def centre_location(self):\n",
        "    \"returns the location of the geometrical centre as a vector\"\n",
        "    return self.geo.mean(axis=0)\n",
        "\n",
        "  def shift(self, displacement):\n",
        "    \"Displaces the molecule as a whole, given a 3D vector (np.array)\"\n",
        "    self.geo += displacement\n",
        "\n",
        "  def center(self):\n",
        "    \"Displaces the molecule so that its geometrical centre is at the origin.\"\n",
        "    self.shift(-self.centre_location())\n",
        "\n",
        "  def write_xyz(self,fn=None):\n",
        "    s = f\"{self.natoms()}\\n Created by Molecule Class\\n\"\n",
        "    for n in range(self.natoms()):\n",
        "        s += f\"{self.symbols[n]:3s}\"\n",
        "        for i in range(3):\n",
        "          s += f\" {self.geo[n,i]:16.6f}\"\n",
        "        s += '\\n'\n",
        "    if fn:\n",
        "      with open(fn,'w') as f:\n",
        "        f.write(s)\n",
        "    else:\n",
        "      return s\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GmH2DuVf0_BW"
      },
      "source": [
        "def shell(cmd, shell=True):\n",
        "  \"runs a command in the linux shell, adapted from Jimmy Kromann.\"\n",
        "  if shell:\n",
        "    p = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "  else:\n",
        "    p = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
        "  output, err = p.communicate()\n",
        "  return output\n",
        "\n",
        "def run_xtb(args):\n",
        "  return shell(f\"ulimit -s unlimited; OMP_STACKSIZE=12G OMP_MAX_ACTIVE_LEVELS=1 /content/xtb-6.5.1/bin/xtb {args}\")\n",
        "\n",
        "def get_snapshot(dirname='/content/work_dir',fn=None):\n",
        "  \"Downloads work_dir as a tbz file for archive.\"\n",
        "  from google.colab import files\n",
        "  import os\n",
        "  import time\n",
        "  origin = os.path.abspath(os.curdir)\n",
        "  now=time.localtime()\n",
        "  date_str=f\"{now[0]-2000:02d}{now[1]:02d}{now[2]:02d}-{now[3]:02d}{now[4]:02d}\"\n",
        "  ofn = fn if fn else f\"{date_str}-{os.path.split(dirname)[-1]}.tbz\"\n",
        "  os.chdir(dirname)\n",
        "  os.chdir('..')\n",
        "  shell(f\"tar cjf {ofn} ./{os.path.split(dirname)[-1]}\")\n",
        "  files.download(ofn)\n",
        "  os.chdir(origin)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CnICNhMr2U0l"
      },
      "source": [
        "# Raios covalentes para os elementos\n",
        "coord_raddii = dict()\n",
        "coord_raddii['H']= 0.5\n",
        "coord_raddii['B']=0.85\n",
        "coord_raddii['C']= 0.8\n",
        "coord_raddii['N']= 0.8\n",
        "coord_raddii['O']= 0.7\n",
        "coord_raddii['Cl']= 1.05\n",
        "\n",
        "atomic_symbols=['XX','H','He','Li','Be','B','C','N','O','F','Ne']\n",
        "\n",
        "def extract_geo(ifn, n, ofn):\n",
        "  \"\"\"Extrai a geometria n de um ficheiro do tipo xtb*.log (começando em 1) e \n",
        "  guarda essa geometria num ficheiro chamod ofn\"\"\"\n",
        "  with open(ifn,'r') as f:\n",
        "    data=f.readlines()\n",
        "  natoms = int(data[0])\n",
        "  start = (natoms+2)*(n-1)\n",
        "  end = start + natoms + 2\n",
        "  with open(ofn,'w') as f:\n",
        "    for line in data[start:end]:\n",
        "      f.write(line)\n",
        "\n",
        "def llm(fn1, fn2, nome, n=10, xtb_args='' ):\n",
        "  \"\"\"Corre o xtb de forma a obter o perfil da LLM entre dois estados,\n",
        "  guardados no ficheiro fn1, e fn2\"\"\"\n",
        "  e1 = Molecule(fn1)\n",
        "  e2 = Molecule(fn2)\n",
        "  energias = list()\n",
        "  for i,k in enumerate(np.linspace(0.0, 1.0, n)):\n",
        "    point = Molecule()\n",
        "    point.symbols=e1.symbols\n",
        "    point.name=f\"LMM point {n}\"\n",
        "    point.geo = (1.0-k)* e1.geo + k*e2.geo\n",
        "    point.write_xyz(f\"{nome}-lmm{i:03d}-guess.xyz\")\n",
        "    run_xtb(f\"{nome}-lmm{i:03d}-guess.xyz --scc --namespace {nome}-lmm{i:03d} {xtb_args} > {nome}-lmm{i:03d}.out\")\n",
        "    with open(f\"{nome}-lmm{i:03d}.out\", 'r') as f:\n",
        "      data = f.readlines()\n",
        "    for line in data:\n",
        "      if ' | TOTAL ENERGY ' in line:\n",
        "        energias.append(float(line.split()[-3]))\n",
        "  for i,e in enumerate(energias):\n",
        "    print(f\"{i:5d} {e:16.5f}\")\n",
        "  plt.plot(energias)\n",
        "  plt.show()\n",
        "\n",
        "def extract_geo_mtd_velo(ifn, n, ofn):\n",
        "  \"\"\"Extrai a geometria n de um ficheiro do tipo xtb.trj (começando em 1) e \n",
        "  guarda essa geometria num ficheiro chamod ofn. Originario de uma dinamica com velocidades.\"\"\"\n",
        "  with open(ifn,'r') as f:\n",
        "    data=f.readlines()\n",
        "  natoms = int(data[0])\n",
        "  start = ((2*natoms)+2)*(n-1)\n",
        "  end = start + natoms + 2\n",
        "  with open(ofn,'w') as f:\n",
        "    for line in data[start:end]:\n",
        "      f.write(line)\n",
        "\n",
        "def g98_animate_mode(fn,n_vib,scale=1.0):\n",
        "  data = open(fn,'r').readlines()\n",
        "  ZZ = list()\n",
        "  geo = list()\n",
        "  displ = list()\n",
        "  # ler a geometria\n",
        "  geo_start=100000000000000\n",
        "  geo_end = -1\n",
        "  for n,line in enumerate(data):\n",
        "    if 'Coordinates' in line:\n",
        "      geo_start = n+3\n",
        "    if (n>geo_start) and '--------' in line:\n",
        "      geo_end = n\n",
        "      break\n",
        "  for n in range(geo_start,geo_end):\n",
        "    l =  data[n].split()\n",
        "    ZZ.append(int(l[1]))\n",
        "    geo.append(l[-3:])\n",
        "  symbols = [atomic_symbols[z] for z in ZZ]\n",
        "  natoms = len(symbols)\n",
        "  vib_start = geo_end + 7\n",
        "  vib_step = 9 + natoms\n",
        "  for i in range(vib_start,len(data),vib_step):\n",
        "    if n_vib in list(map(int,data[i].split())):\n",
        "      v_ind = list(map(int,data[i].split())).index(n_vib)\n",
        "      v_start = i + 9\n",
        "      v_end = i + vib_step\n",
        "      for j in range(v_start,v_end):\n",
        "        l = data[j].split()\n",
        "        displ.append(list(map(float,l[2+(3*v_ind):5+(3*v_ind)])))\n",
        "      break\n",
        "  for n in range(natoms):\n",
        "    for j in range(3):\n",
        "      displ[n][j] *= scale\n",
        "  xyz=f\"{natoms}\\n* (null), Energy   -1000.0000000\\n\"\n",
        "  for n in range(natoms):\n",
        "    xyz += f\"{symbols[n]}\"\n",
        "    for j in range(3):\n",
        "      xyz += f\" {geo[n][j]}\"\n",
        "    for j in range(3):\n",
        "      xyz += f\" {displ[n][j]}\"\n",
        "    xyz += '\\n'\n",
        "  xyzview = py3Dmol.view(width=400,height=400)\n",
        "  xyzview.addModel(xyz,'xyz',{'vibrate': {'frames':10,'amplitude':1}})\n",
        "  xyzview.setStyle({'stick':{},'sphere':{'scale':0.3}})\n",
        "  xyzview.setBackgroundColor('0xeeeeee')\n",
        "  xyzview.animate({'loop': 'backAndForth'})\n",
        "  xyzview.zoomTo()\n",
        "  xyzview.show()\n",
        "\n",
        "def displace_TS(fn,n_vib=1,alpha=1.0,ofn=None,visualize=True):\n",
        "  data = open(fn,'r').readlines()\n",
        "  ZZ = list()\n",
        "  geo = list()\n",
        "  displ = list()\n",
        "  # ler a geometria\n",
        "  geo_start=100000000000000\n",
        "  geo_end = -1\n",
        "  for n,line in enumerate(data):\n",
        "    if 'Coordinates' in line:\n",
        "      geo_start = n+3\n",
        "    if (n>geo_start) and '--------' in line:\n",
        "      geo_end = n\n",
        "      break\n",
        "  for n in range(geo_start,geo_end):\n",
        "    l =  data[n].split()\n",
        "    ZZ.append(int(l[1]))\n",
        "    geo.append(list(map(float,l[-3:])))\n",
        "  symbols = [atomic_symbols[z] for z in ZZ]\n",
        "  natoms = len(symbols)\n",
        "  vib_start = geo_end + 7\n",
        "  vib_step = 9 + natoms\n",
        "  for i in range(vib_start,len(data),vib_step):\n",
        "    if n_vib in list(map(int,data[i].split())):\n",
        "      v_ind = list(map(int,data[i].split())).index(n_vib)\n",
        "      v_start = i + 9\n",
        "      v_end = i + vib_step\n",
        "      for j in range(v_start,v_end):\n",
        "        l = data[j].split()\n",
        "        displ.append(list(map(float,l[2+(3*v_ind):5+(3*v_ind)])))\n",
        "      break\n",
        "  displ=np.array(displ) * alpha\n",
        "  new_geo = np.array(geo) + displ\n",
        "  xyz=f\"{natoms}\\n* (null), Energy   -1000.0000000\\n\"\n",
        "  for n in range(natoms):\n",
        "    xyz += f\"{symbols[n]}\"\n",
        "    for j in range(3):\n",
        "      xyz += f\" {new_geo[n][j]:16.6f}\"\n",
        "    xyz += '\\n'\n",
        "  if ofn:\n",
        "    with open(ofn,'w') as f:\n",
        "      f.write(xyz)\n",
        "  if visualize:\n",
        "    xyzview = py3Dmol.view(width=400,height=400)\n",
        "    xyzview.addModel(xyz,'xyz')\n",
        "    xyzview.setStyle({'stick':{}})\n",
        "    xyzview.setBackgroundColor('0xeeeeee')\n",
        "    xyzview.zoomTo()\n",
        "    xyzview.show()\n",
        "\n",
        "\n",
        "def distances(f, a1, a2):\n",
        "  \"\"\"Devolve a distancia entre o atomo a1 e o atomo a2. Se a1 ou a2 for uma string\n",
        "  devolve todas as distancias que envolvam atomos com esse simbolo.\"\"\"\n",
        "  o = list()\n",
        "  g = f[3] # a geometria é o 3o elemento de uma frame\n",
        "  indx_1 = list()\n",
        "  indx_2 = list()\n",
        "  if type(a1) is int:\n",
        "    indx_1.append(a1-1)\n",
        "  elif type(a1) is str:\n",
        "    for i, s in enumerate(f[2]): # percorrer a lista de simbolos na frame\n",
        "      if s.strip() == a1.strip(): indx_1.append(i)\n",
        "  if type(a2) is int:\n",
        "    indx_2.append(a2-1)\n",
        "  elif type(a2) is str:\n",
        "    for i, s in enumerate(f[2]): # percorrer a lista de simbolos na frame\n",
        "      if s.strip() == a2.strip(): indx_2.append(i)\n",
        "  for i1 in indx_1:\n",
        "    for i2 in indx_2:\n",
        "      d = np.linalg.norm(g[i1]-g[i2])\n",
        "      o.append(d)\n",
        "  return(o)\n",
        "\n",
        "def coord_numbers(f, aidx):\n",
        "  \"\"\"Devolve a distancia entre o atomo a1 e o atomo a2. Se a1 ou a2 for uma string\n",
        "  devolve todas as distancias que envolvam atomos com esse simbolo.\"\"\"\n",
        "  o = list()\n",
        "  g = f[3] # a geometria é o 3o elemento de uma frame\n",
        "  natoms = len(f[2])\n",
        "  indx_1 = list()\n",
        "  if type(aidx) is int:\n",
        "    indx_1.append(aidx-1)\n",
        "  elif type(aidx) is str:\n",
        "    for i, s in enumerate(f[2]): # percorrer a lista de simbolos na frame\n",
        "      if s.strip() == aidx.strip(): indx_1.append(i)\n",
        "  for i1 in indx_1:\n",
        "    c = 0\n",
        "    for i2 in range(natoms):\n",
        "      if i1 == i2:\n",
        "        continue\n",
        "      d = np.linalg.norm(g[i1]-g[i2])\n",
        "      if d <= coord_raddii[f[2][i1]]+coord_raddii[f[2][i2]]:\n",
        "        c += 1\n",
        "    o.append(c)\n",
        "  return(o)\n",
        "\n",
        "def distances_pp(f, ag1, ag2):\n",
        "  \"\"\"Devolve a distancia minima entre o grupo de atomos ag1 e \n",
        "  o grupo de átomos a2. \"\"\"\n",
        "  g = f[3] # a geometria é o 3o elemento de uma frame\n",
        "  o = 140000.0\n",
        "  for i1 in ag1:\n",
        "    for i2 in ag2:\n",
        "      d = np.linalg.norm(g[i1]-g[i2])\n",
        "      if d < o:\n",
        "        o = d\n",
        "  return o\n",
        "\n",
        "def angles_pp(f, ag1, ag2):\n",
        "  \"\"\"Devolve o angulo entre os planos definidos pelos grupos de átomo ag1 e ag2\"\"\"\n",
        "  g = f[3]\n",
        "  va1 = g[ag1[0]]-g[ag1[1]]\n",
        "  vb1 = g[ag1[2]]-g[ag1[1]]\n",
        "  va2 = g[ag2[0]]-g[ag2[1]]\n",
        "  vb2 = g[ag2[2]]-g[ag2[1]]\n",
        "  n1 = np.cross(va1,vb1)\n",
        "  n2 = np.cross(va2,vb2)\n",
        "  n1 /= np.linalg.norm(n1)\n",
        "  n2 /= np.linalg.norm(n2)\n",
        "  atmp=np.dot(n1,n2)/(np.linalg.norm(n1)*np.linalg.norm(n2))\n",
        "  if(atmp>1.0):\n",
        "    phi=np.arccos(1.0)\n",
        "  elif(atmp<-1.0):\n",
        "    phi=np.arccos(-1.0)\n",
        "  else:\n",
        "    phi=np.arccos(atmp)\n",
        "  return np.rad2deg(phi)\n",
        "\n",
        "def analisar_trajectoria(fn, colvars, velo=False):\n",
        "  \"\"\"Analisa uma tracetoria de MD do XTB (ficheiro xtb.trj), \n",
        "  e devolve estatisticas sobre varias coordernadas.\n",
        "  \n",
        "  As coordenadas a estudar são dadas como uma lista, em que cada elemento é ele proprio\n",
        "  uma lista com uma string (D, A, D, C) seguida por um número variável de indices ou simbolos atómicos.\n",
        "  \n",
        "  ex: colvars=[[D,1,2], [D,N,H]]\n",
        "  duas variáveis: distancia do átomo 1 ao atomo 2, distancia entre todos os N's e todos os H's\n",
        "\n",
        "  Return:\n",
        "  dicionario, cujas keys sejam o nome de cada variavel e os valores sejam a coleccao\n",
        "  dos valores dessa variavel ao longo da trajectoria (lista)\n",
        "  \"\"\"\n",
        "  #print(fn)\n",
        "  # abrir a trajectoria e dividir por frames. -> lista de frames\n",
        "  raw_traj = open(fn,'r').readlines()\n",
        "  #print(raw_traj)\n",
        "  natoms = int(raw_traj[0])\n",
        "  trj = list()\n",
        "  step = natoms+2\n",
        "  if velo: step = (2*natoms)+2\n",
        "  for n in range(0,len(raw_traj),step):\n",
        "    #print(n+1)\n",
        "    l = raw_traj[n+1].replace(',',' ').split()\n",
        "    for i, token in enumerate(l):\n",
        "      if 'energy' in token:\n",
        "        energy = float(l[i+1])\n",
        "    #time =float(l[5])\n",
        "    symbols = list()\n",
        "    geo = list()\n",
        "    velo = list()\n",
        "    for j in range(n+2,n+2+natoms):\n",
        "      l = raw_traj[j].split()\n",
        "      symbols.append(l[0])\n",
        "      geo.append(np.array(list(map(float,l[1:4]))))\n",
        "    if velo:\n",
        "      for j in range(n+2+natoms,n+2+(2*natoms)):\n",
        "        try:\n",
        "          l=raw_trajs[j].split()\n",
        "        except:\n",
        "          l=['0.0','0.0','0.0']\n",
        "        velo.append(np.array(list(map(float,l))))\n",
        "    else:\n",
        "      for j in range(natoms):\n",
        "        velo.append(np.zeros(3))\n",
        "    trj.append([energy,time,symbols,geo,velo])\n",
        "  #print(trj)\n",
        "  # iniciar o dicionario onde vamos guardar os valores\n",
        "  o = {}\n",
        "  dpp_count = 0\n",
        "  app_count = 0\n",
        "  for var in colvars:\n",
        "    if var[0].upper() == 'DPP':\n",
        "      dpp_count += 1\n",
        "      o[f'DPP_{dpp_count:02d}']=list()\n",
        "    elif var[0].upper() == 'APP':\n",
        "      app_count += 1\n",
        "      o[f'APP_{app_count:02d}']=list()\n",
        "    else:\n",
        "      o['_'.join(map(str,var))]=list() # o nome da variavel é X_idx1_idx2_etc..\n",
        "  # Para cada frame, aplicar os calculos necessarios para recoher os dados\n",
        "  for frame in trj:\n",
        "    dpp_count = 0\n",
        "    app_count = 0\n",
        "    for var in colvars:\n",
        "      if var[0].upper() == 'D':\n",
        "        #distancia, esperamos dois indices, que podem ser\n",
        "        # numeros (indices do atomo na geometria), ou letras\n",
        "        # (symbolo atomico de um elemento).\n",
        "        # vamos criar uma funcao distances que trata de recolher a(s) distancias\n",
        "        # pretendidas\n",
        "        for val in distances(frame, var[1], var[2]):\n",
        "          o['_'.join(map(str,var))].append(val)\n",
        "      elif var[0].upper() == 'E':\n",
        "        o['E'] = [x[0] for x in trj]\n",
        "      elif var[0].upper() == 'C':\n",
        "        #distancia, esperamos im indices, que pode ser\n",
        "        # numero (indice do atomo na geometria), ou letras\n",
        "        # (symbolo atomico de um elemento).\n",
        "        # vamos criar uma funcao distances que trata de recolher a(s) distancias\n",
        "        # pretendidas\n",
        "        for val in coord_numbers(frame, var[1]):\n",
        "          o['_'.join(map(str,var))].append(val)\n",
        "      elif var[0].upper() == 'DPP':\n",
        "        dpp_count += 1\n",
        "        #distancia, esperamos dois ggrupos de átomos.\n",
        "        o[f'DPP_{dpp_count:02d}'].append(distances_pp(frame,var[1],var[2]))\n",
        "      elif var[0].upper() == 'APP':\n",
        "        app_count += 1\n",
        "        #angulo entre dois planos, esperamos duas triades de átomos.\n",
        "        o[f'APP_{app_count:02d}'].append(angles_pp(frame,var[1],var[2]))\n",
        "  # return os dados\n",
        "  return o\n",
        "\n",
        "def get_closest_geo(fn, colvars, ofn, velo= True):\n",
        "  \"\"\"Dado um ficheiro de trajectória (fn), guarda o ponto cuja geometria mais se aproxima das\n",
        "  consições dadas na lista colvars (semelante ao analisar_trajectoria)\n",
        "    \n",
        "  ex: colvars=[[D, 1, 2, 1.2], [D,3, 4, 3.5]]\n",
        "  duas variáveis: escolhe o ponto mais próximo de ter a distancia 1-2 a 1.2angs, e a distancia 3-4 a 3.5 angs.\n",
        "  \n",
        "  no ficheiro ofn.\"\"\"\n",
        "  vars = analisar_trajectoria(fn,[x[:-1 ] for x in colvars], velo= velo)\n",
        "  var_names = list(vars.keys())\n",
        "  objectivos = dict()\n",
        "  for var in colvars:\n",
        "    objectivos['_'.join(map(str,var[:-1]))]=var[-1] # o nome da variavel é X_idx1_idx2_etc.. e o objectivo é o ultimo valor\n",
        "  n = 0\n",
        "  best_n = 0\n",
        "  best_score = np.sum([ np.abs(vars[vn][n]-objectivos[vn]) for vn in var_names])\n",
        "  while n < len(vars[var_names[0]]):\n",
        "    score = np.sum([ np.abs(vars[vn][n]-objectivos[vn]) for vn in var_names])\n",
        "    if score < best_score:\n",
        "      best_n = n\n",
        "      best_score = score\n",
        "    n += 1\n",
        "  if velo:\n",
        "    extract_geo_mtd_velo(fn,best_n+1,ofn)\n",
        "  else:\n",
        "    extract_geo(fn,best_n+1,ofn)\n",
        "\n",
        "def free_energy_path_1D(CV, cv_name='Collective Variable', T= 298.15, plot= True, npoints=128, ylim=None):\n",
        "  cv = np.array(CV)\n",
        "  if ylim:\n",
        "    tmp_y = np.linspace(ylim[0],ylim[1],npoints)\n",
        "  else:\n",
        "    tmp_y = np.linspace(0.9*min(cv),1.1*max(CV),npoints)\n",
        "  hist, vals = np.histogram(cv,bins=tmp_y)\n",
        "  DeltaG = -8.314e-3 * T * np.log(hist/hist.max())\n",
        "  if plot:\n",
        "    fig, ax = plt.subplots(figsize=(18,9))\n",
        "    ax.plot(vals[:-1],DeltaG)\n",
        "    ax.ticklabel_format(axis='both', useOffset=False)\n",
        "    ax.set_ylabel(\"Free Energy / kJ/mol\")\n",
        "    ax.set_xlabel(f\"{cv_name}\")\n",
        "    plt.show()\n",
        "  return (DeltaG, vals[:-1])\n",
        "\n",
        "def free_energy_path_2D(CV1, CV2, cv_name1='Collective Variable 1', cv_name2=\"Collective Variable 2\", T= 298.15, plot= True, npoints=128, ylim1=None, ylim2=None):\n",
        "  cv1 = np.array(CV1)\n",
        "  cv2 = np.array(CV2)\n",
        "  if ylim1:\n",
        "    tmp_y1 = np.linspace(ylim1[0],ylim1[1],npoints)\n",
        "  else:\n",
        "    tmp_y1 = np.linspace(0.9*min(cv1),1.1*max(cv1),npoints)\n",
        "  if ylim2:\n",
        "    tmp_y2 = np.linspace(ylim2[0],ylim2[1],npoints)\n",
        "  else:\n",
        "    tmp_y2 = np.linspace(0.9*min(cv2),1.1*max(cv2),npoints)\n",
        "  hist, vals_x, vals_y = np.histogram2d(cv1, cv2,bins=(tmp_y1,tmp_y2))\n",
        "  DeltaG = -8.314e-3 * T * np.log(hist/hist.max())\n",
        "  if plot:\n",
        "    fig, ax = plt.subplots(figsize=(12,10))\n",
        "    X, Y = np.meshgrid(vals_x[:-1],vals_y[:-1])\n",
        "    cs = ax.contourf(X,Y,DeltaG)\n",
        "    cbar = fig.colorbar(cs)\n",
        "    ax.set_ylabel(f\"{cv_name2}\")\n",
        "    ax.set_xlabel(f\"{cv_name1}\")\n",
        "    plt.show()\n",
        "  return (DeltaG, X, Y)\n",
        "\n",
        "def grafico_energia_path(fn):\n",
        "  \"\"\"Dá o grafico com a variação da energia ao longo da trajectoria no fincheiro fn\"\"\"\n",
        "  # abrir a trajectoria e dividir por frames. -> lista de frames\n",
        "  raw_traj = open(fn,'r').readlines()\n",
        "  #print(raw_traj)\n",
        "  natoms = int(raw_traj[0])\n",
        "  trj = list()\n",
        "  for n in range(0,len(raw_traj),natoms+2):\n",
        "    #print(n+1)\n",
        "    l = raw_traj[n+1].split()\n",
        "    energy = float(l[1])\n",
        "    symbols = list()\n",
        "    geo = list()\n",
        "    for j in range(n+2,n+2+natoms):\n",
        "      l = raw_traj[j].split()\n",
        "      symbols.append(l[0])\n",
        "      geo.append(np.array(list(map(float,l[1:4]))))\n",
        "    trj.append([energy,symbols,geo])\n",
        "  energies = [x[0] for x in trj]\n",
        "  steps = np.arange(len(energies))\n",
        "  fig, ax = plt.subplots()\n",
        "  ax.plot(steps,energies)\n",
        "  ax.set_ylabel('Total Energy / Eh')\n",
        "  ax.set_xlabel('MD Step')\n",
        "  ax.ticklabel_format(axis='both', useOffset=False)\n",
        "  plt.show()\n",
        "\n",
        "### Stuff added to ensure atom order in the reactants and products is as consistent as possible for --path calculations\n",
        "\n",
        "def read_xyz(fn):\n",
        "    data = open(fn,'r').readlines()\n",
        "    natoms = int(data[0])\n",
        "    symbols = list()\n",
        "    geo=np.zeros((natoms,3))\n",
        "    for i in range(2,2+natoms):\n",
        "        l = data[i].split()\n",
        "        symbols.append(l[0])\n",
        "        geo[i-2,:]=list(map(float,l[1:4]))\n",
        "    return (symbols,geo)\n",
        "\n",
        "def write_xyz(geo, fn):\n",
        "    with open(fn,'w') as f:\n",
        "        f.write(f\"{len(geo[0]):6d}\\nWriten by rearranje atoms\\n\")\n",
        "        for i,s in enumerate(geo[0]):\n",
        "            f.write(f\" {s:2s}\")\n",
        "            for j in range(3):\n",
        "                f.write(f\" {geo[1][i,j]:16.6f}\")\n",
        "            f.write('\\n')\n",
        "\n",
        "def make_distance_matrix(geo):\n",
        "    s = geo[0]\n",
        "    g = geo[1]\n",
        "    natoms=len(geo[0])\n",
        "    dm=np.zeros((natoms,natoms))\n",
        "    for i in range(natoms-1):\n",
        "        for j in range(i+1,natoms):\n",
        "            d = np.linalg.norm(g[j,:]-g[i,:])\n",
        "            dm[i,j]=d\n",
        "            dm[j,i]=d\n",
        "    return dm\n",
        "    \n",
        "def atom_exchange(geo,elements='all'):\n",
        "    #makes 1 permutation within atoms of the same element,\n",
        "    # for each element in elements (list, or 'all')\n",
        "    do_elements=list(set(geo[0])) # all elements in geo\n",
        "    if type(elements) is list:\n",
        "        do_elements = elements\n",
        "    new_g = geo[1].copy()\n",
        "    for element in do_elements:\n",
        "        if geo[0].count(element) <= 1:\n",
        "            continue\n",
        "        # find indexes of atoms with element\n",
        "        idxs = np.where(np.array(geo[0])==element)[0]\n",
        "        # make a permutation\n",
        "        new_idxs = np.random.permutation(idxs)\n",
        "        for i in range(len(idxs)):\n",
        "            new_g[new_idxs[i],:]=geo[1][idxs[i],:]\n",
        "    # return the mutated geometry onto a new tuple\n",
        "    return (geo[0],new_g)\n",
        "\n",
        "def rearrange_atoms(geo_fn, ref_fn, max_iter=10000, accept_ratio = 0.05):\n",
        "    geo0 = read_xyz(geo_fn)\n",
        "    ref  = read_xyz(ref_fn)\n",
        "    #TODO: check is symbols are in same order\n",
        "    ref_dm = make_distance_matrix(ref)\n",
        "    # simple MC opotimization\n",
        "    n_iter = 0\n",
        "    global_best = geo0\n",
        "    global_best_score = np.sum((make_distance_matrix(geo0)-ref_dm)**2)\n",
        "    last = geo0\n",
        "    last_score = global_best_score\n",
        "    while n_iter <= max_iter:\n",
        "        n_iter += 1\n",
        "        #s = f\"{n_iter:5d}\"\n",
        "        probe=atom_exchange(last)\n",
        "        probe_score = np.sum((make_distance_matrix(probe)-ref_dm)**2)\n",
        "        if probe_score < global_best_score:\n",
        "            global_best = probe\n",
        "            global_best_score = probe_score\n",
        "            last = probe\n",
        "            last_score = probe_score\n",
        "        elif probe_score < last_score:\n",
        "            last = probe\n",
        "            last_score = probe_score\n",
        "        elif probe_score > last_score:\n",
        "            x = np.random.random_sample()\n",
        "            if x <= accept_ratio:\n",
        "                last = probe\n",
        "                last_score = probe_score\n",
        "        #s += f\" {global_best_score:10.3f} {last_score:10.3f} {probe_score:10.3f}\"\n",
        "        #print(s)\n",
        "        del(probe)\n",
        "    return global_best\n",
        "\n",
        "def ener2pandas(fn):\n",
        "  with open(fn,'r') as f:\n",
        "    idata=f.readlines()\n",
        "  o=dict()\n",
        "  header=['Step','Time_fs','EKin_au','Temp_K','EPot_au','ConsQty_au','TimeUsed_s']\n",
        "  for cn in header:\n",
        "    o[cn]=list()\n",
        "  for line in idata[1:]:\n",
        "    l=line.split()\n",
        "    for i,item in enumerate(l):\n",
        "      o[header[i]].append(float(item))\n",
        "  o=pd.DataFrame(o)\n",
        "  return(o)\n",
        "\n",
        "def ener2csv(in_fn, csv_fn):\n",
        "  ener2pandas(in_fn).to_csv(csv_fn, index=False)\n",
        "\n",
        "def compile_reaction_data(r_fn, ts_fn, p_fn):\n",
        "  r={'E':0.0,'H':0.0,'G':0.0}\n",
        "  ts={'E':0.0,'H':0.0,'G':0.0,'Ivib':0.0}\n",
        "  p={'E':0.0,'H':0.0,'G':0.0}\n",
        "  # reactant\n",
        "  data=open(r_fn).readlines()\n",
        "  for line in data[-100:]:\n",
        "    if '| TOTAL ENERGY ' in line:\n",
        "      r['E']=float(line.split()[-3])\n",
        "    if '| TOTAL ENTHALPY' in line:\n",
        "      r['H']=float(line.split()[-3])\n",
        "    if 'TOTAL FREE ENERGY' in line:\n",
        "      r['G']=float(line.split()[-3])\n",
        "  # TS\n",
        "  data=open(ts_fn).readlines()\n",
        "  for line in data[-100:]:\n",
        "    if '| TOTAL ENERGY ' in line:\n",
        "      ts['E']=float(line.split()[-3])\n",
        "    if '| TOTAL ENTHALPY' in line:\n",
        "      ts['H']=float(line.split()[-3])\n",
        "    if 'TOTAL FREE ENERGY' in line:\n",
        "      ts['G']=float(line.split()[-3])\n",
        "  data = open(ts_fn.replace('.out','.vibspectrum'),'r').readlines()\n",
        "  for line in data[4:]:\n",
        "    try:\n",
        "      freq=float(line[24:38].strip())\n",
        "    except:\n",
        "      freq=1.0\n",
        "    if freq < -1.0:\n",
        "      ts['Ivib']=freq\n",
        "      break\n",
        "  # product\n",
        "  data=open(p_fn).readlines()\n",
        "  for line in data[-100:]:\n",
        "    if '| TOTAL ENERGY ' in line:\n",
        "      p['E']=float(line.split()[-3])\n",
        "    if '| TOTAL ENTHALPY' in line:\n",
        "      p['H']=float(line.split()[-3])\n",
        "    if 'TOTAL FREE ENERGY' in line:\n",
        "      p['G']=float(line.split()[-3])\n",
        "  reactant = Molecule(r_fn.replace(\".out\",'.xtbopt.xyz'))\n",
        "  reactant.show()\n",
        "  print(\"Reactant final geometry\")\n",
        "  g98_animate_mode(ts_fn.replace('.out','.g98.out'),1,1.0)\n",
        "  print(\"Animation of the Reaction Coord. at the TS\")\n",
        "  product = Molecule(p_fn.replace(\".out\",'.xtbopt.xyz'))\n",
        "  product.show()\n",
        "  print(\"Reactant final geometry\")\n",
        "  Eh2kjmol = 2625.5 # conversao Eh -> kJ/mol\n",
        "  de_d = (p['E']-r['E']) * Eh2kjmol\n",
        "  dh_d = (p['H']-r['H']) * Eh2kjmol\n",
        "  dg_d = (p['G']-r['G']) * Eh2kjmol\n",
        "  ea_d = (ts['E']-r['E']) * Eh2kjmol\n",
        "  ha_d = (ts['H']-r['H']) * Eh2kjmol\n",
        "  ga_d = (ts['G']-r['G']) * Eh2kjmol\n",
        "  de_i = (r['E']-p['E']) * Eh2kjmol\n",
        "  dh_i = (r['H']-p['H']) * Eh2kjmol\n",
        "  dg_i = (r['G']-p['G']) * Eh2kjmol\n",
        "  ea_i = (ts['E']-p['E']) * Eh2kjmol\n",
        "  ha_i = (ts['H']-p['H']) * Eh2kjmol\n",
        "  ga_i = (ts['G']-p['G']) * Eh2kjmol\n",
        "  print(f\"           {'Energy':^16s} {'Enthalpy':^16s} {'Gibbs Energy':^16s}\")\n",
        "  print(f\" Delta (D) {de_d:16.2f} {dh_d:16.2f} {dg_d:16.2f}\")\n",
        "  print(f\" Eact (D)  {ea_d:16.2f} {ha_d:16.2f} {ha_d:16.2f}\")\n",
        "  print(f\" Delta (I) {de_i:16.2f} {dh_i:16.2f} {dg_i:16.2f}\")\n",
        "  print(f\" Eact (I)  {ea_i:16.2f} {ha_i:16.2f} {ha_i:16.2f}\")\n",
        "\n",
        "  print(f\"Value of the imaginary frequency at TS: {ts['Ivib']:20.2f}\")\n",
        "\n",
        "def calc_bond_matrix(g, radii):\n",
        "  natoms = g.shape[0]\n",
        "  o = np.zeros((natoms,natoms),dtype=np.int32)\n",
        "  for i in range(natoms-1):\n",
        "    for j in range(i+1,natoms):\n",
        "      if np.linalg.norm(g[i]-g[j]) <= (radii[i]+radii[j]) :\n",
        "        o[i,j] = 1\n",
        "  return o\n",
        "\n",
        "def formula_from_bond_matrix(bond_matrix, symbols, i_ref):\n",
        "  d = dict()\n",
        "  stack=[i_ref]\n",
        "  visited=list()\n",
        "  natoms = len(symbols)\n",
        "  while len(stack)>0:\n",
        "    j = stack.pop()\n",
        "    s = symbols[j]\n",
        "    visited.append(j)\n",
        "    if s not in d:\n",
        "      d[s] = 1\n",
        "    else:\n",
        "      d[s] += 1\n",
        "    for i in range(natoms):\n",
        "      if bond_matrix[i,j]==1 or bond_matrix[j,i]==1:\n",
        "        if (i not in visited) and (i not in stack):\n",
        "          stack.append(i)\n",
        "  # compose text\n",
        "  o=''\n",
        "  for symbol in ['C','H','N','P','O','S','F','Cl']:\n",
        "    if symbol in d:\n",
        "      o += symbol\n",
        "      if d[symbol] > 1:\n",
        "        o += f\"{d[symbol]}\"\n",
        "  visited.sort()\n",
        "  return (o, visited)\n",
        "\n",
        "def isolate_molecules_with_atoms(fn, atom_lst, ofn):\n",
        "  \"\"\"Lê o ficheiro xyz fn e isola as moléculas que contém os átomos cujos indices\n",
        "  estão em atom_lst. Guarda o resultado no ficheiro ofn\"\"\"\n",
        "  original = Molecule(fn)\n",
        "  o=Molecule()\n",
        "  cov_radii = list()\n",
        "  symbols = original.symbols\n",
        "  for s in symbols:\n",
        "    if s in coord_raddii:\n",
        "      cov_radii.append(coord_raddii[s])\n",
        "    else:\n",
        "      print(f\"No covalent radiu for {s}\")\n",
        "      cov_radii.append(0.76)\n",
        "  bm = calc_bond_matrix(original.geo,cov_radii)\n",
        "  indexes=list()\n",
        "  for a_i in atom_lst:\n",
        "    _, atoms = formula_from_bond_matrix(bm,symbols,a_i)\n",
        "    for a_j in atoms:\n",
        "      if a_j not in indexes:\n",
        "        indexes.append(a_j)\n",
        "  indexes.sort()\n",
        "  o.geo=np.zeros((len(indexes),3))\n",
        "  for i,idx in enumerate(indexes):\n",
        "    o.symbols.append(original.symbols[idx])\n",
        "    o.geo[i,:]=original.geo[idx,:]\n",
        "  o.center()\n",
        "  o.write_xyz(ofn)\n",
        "\n",
        "def historia_trajectoria(fn, velo=False):\n",
        "  \"\"\"Analisa uma tracetoria de MD do XTB (ficheiro xtb.trj), \n",
        "  e devolve histórico das transformações química.\n",
        "  \n",
        "  Return:\n",
        "  dicionario, cujas keys sejam o nome de cada variavel e os valores sejam a coleccao\n",
        "  dos valores dessa variavel ao longo da trajectoria (lista)\n",
        "  \"\"\"\n",
        "  print(fn)\n",
        "  # abrir a trajectoria e dividir por frames. -> lista de frames\n",
        "  raw_traj = open(fn,'r').readlines()\n",
        "  #print(raw_traj)\n",
        "  natoms = int(raw_traj[0])\n",
        "  trj = list()\n",
        "  step = natoms+2\n",
        "  if velo: step = (2*natoms)+2\n",
        "  for n in range(0,len(raw_traj),step):\n",
        "    #print(n+1)\n",
        "    l = raw_traj[n+1].replace(',',' ').split()\n",
        "    for i, token in enumerate(l):\n",
        "      if 'energy' in token:\n",
        "        energy = float(l[i+1])\n",
        "    #time =float(l[5])\n",
        "    symbols = list()\n",
        "    geo = list()\n",
        "    velo = list()\n",
        "    for j in range(n+2,n+2+natoms):\n",
        "      l = raw_traj[j].split()\n",
        "      symbols.append(l[0])\n",
        "      geo.append(np.array(list(map(float,l[1:4]))))\n",
        "    geo = np.array(geo)\n",
        "    if velo:\n",
        "      for j in range(n+2+natoms,n+2+(2*natoms)):\n",
        "        try:\n",
        "          l=raw_trajs[j].split()\n",
        "        except:\n",
        "          l=['0.0','0.0','0.0']\n",
        "        velo.append(np.array(list(map(float,l))))\n",
        "    else:\n",
        "      for j in range(natoms):\n",
        "        velo.append(np.zeros(3))\n",
        "    trj.append([energy,time,symbols,geo,velo])\n",
        "  #print(trj)\n",
        "  cov_radii = list()\n",
        "  symbols = trj[0][2]\n",
        "  for s in trj[0][2]:\n",
        "    if s in coord_raddii:\n",
        "      cov_radii.append(coord_raddii[s])\n",
        "    else:\n",
        "      print(f\"No covalent radiu for {s}\")\n",
        "      cov_radii.append(0.76)\n",
        "  bm0 = calc_bond_matrix(trj[0][3], radii = np.array(cov_radii))\n",
        "  for iframe,frame in enumerate(trj):\n",
        "    bm = calc_bond_matrix(frame[3], radii = np.array(cov_radii))\n",
        "    s=''\n",
        "    changes = False\n",
        "    for i in range(natoms-1):\n",
        "      for j in range(i,natoms):\n",
        "        delta_b = bm[i,j]-bm0[i,j]\n",
        "        if delta_b>0:\n",
        "          changes = True\n",
        "          s += f\"\\n + Formação ligação {symbols[i]} ({i+1}) - {symbols[j]} ({j+1}) (d = {np.linalg.norm(frame[3][i]-frame[3][j]):6.3f} angs)\"\n",
        "          r1, r1i = formula_from_bond_matrix(bm0,symbols,i)\n",
        "          r2, r2i = formula_from_bond_matrix(bm0,symbols,j)\n",
        "          p, pi = formula_from_bond_matrix(bm,symbols,i)\n",
        "          if r1i == r2i:\n",
        "            s += f\"\\n   {r1} -> {p} (isomerização?)\"\n",
        "          else:\n",
        "            s += f\"\\n   {r1} + {r2} -> {p}\"\n",
        "        elif delta_b<0:\n",
        "          changes = True\n",
        "          s += f\"\\n - Quebra de ligação {symbols[i]} ({i+1}) - {symbols[j]} ({j+1}) (d = {np.linalg.norm(frame[3][i]-frame[3][j]):6.3f} angs)\"\n",
        "          r,ri = formula_from_bond_matrix(bm0,symbols,i)\n",
        "          p1,p1i = formula_from_bond_matrix(bm,symbols,i)\n",
        "          p2,p2i = formula_from_bond_matrix(bm,symbols,j)\n",
        "          if p1i == p2i:\n",
        "            s += f\"\\n   {r} -> {p1} (isomerização?)\"\n",
        "          else:\n",
        "            s += f\"\\n   {r} -> {p1} + {p2}\"\n",
        "    if changes:\n",
        "      print(f\"Frame {iframe+1}:{s}\")\n",
        "    bm0 = np.copy(bm)\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Análise"
      ],
      "metadata": {
        "id": "Kcp5pkt9OsgN"
      }
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "G5I6Pn6_Eiom",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "resources": {
            "http://localhost:8080/nbextensions/google.colab/files.js": {
              "data": "Ly8gQ29weXJpZ2h0IDIwMTcgR29vZ2xlIExMQwovLwovLyBMaWNlbnNlZCB1bmRlciB0aGUgQXBhY2hlIExpY2Vuc2UsIFZlcnNpb24gMi4wICh0aGUgIkxpY2Vuc2UiKTsKLy8geW91IG1heSBub3QgdXNlIHRoaXMgZmlsZSBleGNlcHQgaW4gY29tcGxpYW5jZSB3aXRoIHRoZSBMaWNlbnNlLgovLyBZb3UgbWF5IG9idGFpbiBhIGNvcHkgb2YgdGhlIExpY2Vuc2UgYXQKLy8KLy8gICAgICBodHRwOi8vd3d3LmFwYWNoZS5vcmcvbGljZW5zZXMvTElDRU5TRS0yLjAKLy8KLy8gVW5sZXNzIHJlcXVpcmVkIGJ5IGFwcGxpY2FibGUgbGF3IG9yIGFncmVlZCB0byBpbiB3cml0aW5nLCBzb2Z0d2FyZQovLyBkaXN0cmlidXRlZCB1bmRlciB0aGUgTGljZW5zZSBpcyBkaXN0cmlidXRlZCBvbiBhbiAiQVMgSVMiIEJBU0lTLAovLyBXSVRIT1VUIFdBUlJBTlRJRVMgT1IgQ09ORElUSU9OUyBPRiBBTlkgS0lORCwgZWl0aGVyIGV4cHJlc3Mgb3IgaW1wbGllZC4KLy8gU2VlIHRoZSBMaWNlbnNlIGZvciB0aGUgc3BlY2lmaWMgbGFuZ3VhZ2UgZ292ZXJuaW5nIHBlcm1pc3Npb25zIGFuZAovLyBsaW1pdGF0aW9ucyB1bmRlciB0aGUgTGljZW5zZS4KCi8qKgogKiBAZmlsZW92ZXJ2aWV3IEhlbHBlcnMgZm9yIGdvb2dsZS5jb2xhYiBQeXRob24gbW9kdWxlLgogKi8KKGZ1bmN0aW9uKHNjb3BlKSB7CmZ1bmN0aW9uIHNwYW4odGV4dCwgc3R5bGVBdHRyaWJ1dGVzID0ge30pIHsKICBjb25zdCBlbGVtZW50ID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnc3BhbicpOwogIGVsZW1lbnQudGV4dENvbnRlbnQgPSB0ZXh0OwogIGZvciAoY29uc3Qga2V5IG9mIE9iamVjdC5rZXlzKHN0eWxlQXR0cmlidXRlcykpIHsKICAgIGVsZW1lbnQuc3R5bGVba2V5XSA9IHN0eWxlQXR0cmlidXRlc1trZXldOwogIH0KICByZXR1cm4gZWxlbWVudDsKfQoKLy8gTWF4IG51bWJlciBvZiBieXRlcyB3aGljaCB3aWxsIGJlIHVwbG9hZGVkIGF0IGEgdGltZS4KY29uc3QgTUFYX1BBWUxPQURfU0laRSA9IDEwMCAqIDEwMjQ7CgpmdW5jdGlvbiBfdXBsb2FkRmlsZXMoaW5wdXRJZCwgb3V0cHV0SWQpIHsKICBjb25zdCBzdGVwcyA9IHVwbG9hZEZpbGVzU3RlcChpbnB1dElkLCBvdXRwdXRJZCk7CiAgY29uc3Qgb3V0cHV0RWxlbWVudCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKG91dHB1dElkKTsKICAvLyBDYWNoZSBzdGVwcyBvbiB0aGUgb3V0cHV0RWxlbWVudCB0byBtYWtlIGl0IGF2YWlsYWJsZSBmb3IgdGhlIG5leHQgY2FsbAogIC8vIHRvIHVwbG9hZEZpbGVzQ29udGludWUgZnJvbSBQeXRob24uCiAgb3V0cHV0RWxlbWVudC5zdGVwcyA9IHN0ZXBzOwoKICByZXR1cm4gX3VwbG9hZEZpbGVzQ29udGludWUob3V0cHV0SWQpOwp9CgovLyBUaGlzIGlzIHJvdWdobHkgYW4gYXN5bmMgZ2VuZXJhdG9yIChub3Qgc3VwcG9ydGVkIGluIHRoZSBicm93c2VyIHlldCksCi8vIHdoZXJlIHRoZXJlIGFyZSBtdWx0aXBsZSBhc3luY2hyb25vdXMgc3RlcHMgYW5kIHRoZSBQeXRob24gc2lkZSBpcyBnb2luZwovLyB0byBwb2xsIGZvciBjb21wbGV0aW9uIG9mIGVhY2ggc3RlcC4KLy8gVGhpcyB1c2VzIGEgUHJvbWlzZSB0byBibG9jayB0aGUgcHl0aG9uIHNpZGUgb24gY29tcGxldGlvbiBvZiBlYWNoIHN0ZXAsCi8vIHRoZW4gcGFzc2VzIHRoZSByZXN1bHQgb2YgdGhlIHByZXZpb3VzIHN0ZXAgYXMgdGhlIGlucHV0IHRvIHRoZSBuZXh0IHN0ZXAuCmZ1bmN0aW9uIF91cGxvYWRGaWxlc0NvbnRpbnVlKG91dHB1dElkKSB7CiAgY29uc3Qgb3V0cHV0RWxlbWVudCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKG91dHB1dElkKTsKICBjb25zdCBzdGVwcyA9IG91dHB1dEVsZW1lbnQuc3RlcHM7CgogIGNvbnN0IG5leHQgPSBzdGVwcy5uZXh0KG91dHB1dEVsZW1lbnQubGFzdFByb21pc2VWYWx1ZSk7CiAgcmV0dXJuIFByb21pc2UucmVzb2x2ZShuZXh0LnZhbHVlLnByb21pc2UpLnRoZW4oKHZhbHVlKSA9PiB7CiAgICAvLyBDYWNoZSB0aGUgbGFzdCBwcm9taXNlIHZhbHVlIHRvIG1ha2UgaXQgYXZhaWxhYmxlIHRvIHRoZSBuZXh0CiAgICAvLyBzdGVwIG9mIHRoZSBnZW5lcmF0b3IuCiAgICBvdXRwdXRFbGVtZW50Lmxhc3RQcm9taXNlVmFsdWUgPSB2YWx1ZTsKICAgIHJldHVybiBuZXh0LnZhbHVlLnJlc3BvbnNlOwogIH0pOwp9CgovKioKICogR2VuZXJhdG9yIGZ1bmN0aW9uIHdoaWNoIGlzIGNhbGxlZCBiZXR3ZWVuIGVhY2ggYXN5bmMgc3RlcCBvZiB0aGUgdXBsb2FkCiAqIHByb2Nlc3MuCiAqIEBwYXJhbSB7c3RyaW5nfSBpbnB1dElkIEVsZW1lbnQgSUQgb2YgdGhlIGlucHV0IGZpbGUgcGlja2VyIGVsZW1lbnQuCiAqIEBwYXJhbSB7c3RyaW5nfSBvdXRwdXRJZCBFbGVtZW50IElEIG9mIHRoZSBvdXRwdXQgZGlzcGxheS4KICogQHJldHVybiB7IUl0ZXJhYmxlPCFPYmplY3Q+fSBJdGVyYWJsZSBvZiBuZXh0IHN0ZXBzLgogKi8KZnVuY3Rpb24qIHVwbG9hZEZpbGVzU3RlcChpbnB1dElkLCBvdXRwdXRJZCkgewogIGNvbnN0IGlucHV0RWxlbWVudCA9IGRvY3VtZW50LmdldEVsZW1lbnRCeUlkKGlucHV0SWQpOwogIGlucHV0RWxlbWVudC5kaXNhYmxlZCA9IGZhbHNlOwoKICBjb25zdCBvdXRwdXRFbGVtZW50ID0gZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQob3V0cHV0SWQpOwogIG91dHB1dEVsZW1lbnQuaW5uZXJIVE1MID0gJyc7CgogIGNvbnN0IHBpY2tlZFByb21pc2UgPSBuZXcgUHJvbWlzZSgocmVzb2x2ZSkgPT4gewogICAgaW5wdXRFbGVtZW50LmFkZEV2ZW50TGlzdGVuZXIoJ2NoYW5nZScsIChlKSA9PiB7CiAgICAgIHJlc29sdmUoZS50YXJnZXQuZmlsZXMpOwogICAgfSk7CiAgfSk7CgogIGNvbnN0IGNhbmNlbCA9IGRvY3VtZW50LmNyZWF0ZUVsZW1lbnQoJ2J1dHRvbicpOwogIGlucHV0RWxlbWVudC5wYXJlbnRFbGVtZW50LmFwcGVuZENoaWxkKGNhbmNlbCk7CiAgY2FuY2VsLnRleHRDb250ZW50ID0gJ0NhbmNlbCB1cGxvYWQnOwogIGNvbnN0IGNhbmNlbFByb21pc2UgPSBuZXcgUHJvbWlzZSgocmVzb2x2ZSkgPT4gewogICAgY2FuY2VsLm9uY2xpY2sgPSAoKSA9PiB7CiAgICAgIHJlc29sdmUobnVsbCk7CiAgICB9OwogIH0pOwoKICAvLyBXYWl0IGZvciB0aGUgdXNlciB0byBwaWNrIHRoZSBmaWxlcy4KICBjb25zdCBmaWxlcyA9IHlpZWxkIHsKICAgIHByb21pc2U6IFByb21pc2UucmFjZShbcGlja2VkUHJvbWlzZSwgY2FuY2VsUHJvbWlzZV0pLAogICAgcmVzcG9uc2U6IHsKICAgICAgYWN0aW9uOiAnc3RhcnRpbmcnLAogICAgfQogIH07CgogIGNhbmNlbC5yZW1vdmUoKTsKCiAgLy8gRGlzYWJsZSB0aGUgaW5wdXQgZWxlbWVudCBzaW5jZSBmdXJ0aGVyIHBpY2tzIGFyZSBub3QgYWxsb3dlZC4KICBpbnB1dEVsZW1lbnQuZGlzYWJsZWQgPSB0cnVlOwoKICBpZiAoIWZpbGVzKSB7CiAgICByZXR1cm4gewogICAgICByZXNwb25zZTogewogICAgICAgIGFjdGlvbjogJ2NvbXBsZXRlJywKICAgICAgfQogICAgfTsKICB9CgogIGZvciAoY29uc3QgZmlsZSBvZiBmaWxlcykgewogICAgY29uc3QgbGkgPSBkb2N1bWVudC5jcmVhdGVFbGVtZW50KCdsaScpOwogICAgbGkuYXBwZW5kKHNwYW4oZmlsZS5uYW1lLCB7Zm9udFdlaWdodDogJ2JvbGQnfSkpOwogICAgbGkuYXBwZW5kKHNwYW4oCiAgICAgICAgYCgke2ZpbGUudHlwZSB8fCAnbi9hJ30pIC0gJHtmaWxlLnNpemV9IGJ5dGVzLCBgICsKICAgICAgICBgbGFzdCBtb2RpZmllZDogJHsKICAgICAgICAgICAgZmlsZS5sYXN0TW9kaWZpZWREYXRlID8gZmlsZS5sYXN0TW9kaWZpZWREYXRlLnRvTG9jYWxlRGF0ZVN0cmluZygpIDoKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgJ24vYSd9IC0gYCkpOwogICAgY29uc3QgcGVyY2VudCA9IHNwYW4oJzAlIGRvbmUnKTsKICAgIGxpLmFwcGVuZENoaWxkKHBlcmNlbnQpOwoKICAgIG91dHB1dEVsZW1lbnQuYXBwZW5kQ2hpbGQobGkpOwoKICAgIGNvbnN0IGZpbGVEYXRhUHJvbWlzZSA9IG5ldyBQcm9taXNlKChyZXNvbHZlKSA9PiB7CiAgICAgIGNvbnN0IHJlYWRlciA9IG5ldyBGaWxlUmVhZGVyKCk7CiAgICAgIHJlYWRlci5vbmxvYWQgPSAoZSkgPT4gewogICAgICAgIHJlc29sdmUoZS50YXJnZXQucmVzdWx0KTsKICAgICAgfTsKICAgICAgcmVhZGVyLnJlYWRBc0FycmF5QnVmZmVyKGZpbGUpOwogICAgfSk7CiAgICAvLyBXYWl0IGZvciB0aGUgZGF0YSB0byBiZSByZWFkeS4KICAgIGxldCBmaWxlRGF0YSA9IHlpZWxkIHsKICAgICAgcHJvbWlzZTogZmlsZURhdGFQcm9taXNlLAogICAgICByZXNwb25zZTogewogICAgICAgIGFjdGlvbjogJ2NvbnRpbnVlJywKICAgICAgfQogICAgfTsKCiAgICAvLyBVc2UgYSBjaHVua2VkIHNlbmRpbmcgdG8gYXZvaWQgbWVzc2FnZSBzaXplIGxpbWl0cy4gU2VlIGIvNjIxMTU2NjAuCiAgICBsZXQgcG9zaXRpb24gPSAwOwogICAgZG8gewogICAgICBjb25zdCBsZW5ndGggPSBNYXRoLm1pbihmaWxlRGF0YS5ieXRlTGVuZ3RoIC0gcG9zaXRpb24sIE1BWF9QQVlMT0FEX1NJWkUpOwogICAgICBjb25zdCBjaHVuayA9IG5ldyBVaW50OEFycmF5KGZpbGVEYXRhLCBwb3NpdGlvbiwgbGVuZ3RoKTsKICAgICAgcG9zaXRpb24gKz0gbGVuZ3RoOwoKICAgICAgY29uc3QgYmFzZTY0ID0gYnRvYShTdHJpbmcuZnJvbUNoYXJDb2RlLmFwcGx5KG51bGwsIGNodW5rKSk7CiAgICAgIHlpZWxkIHsKICAgICAgICByZXNwb25zZTogewogICAgICAgICAgYWN0aW9uOiAnYXBwZW5kJywKICAgICAgICAgIGZpbGU6IGZpbGUubmFtZSwKICAgICAgICAgIGRhdGE6IGJhc2U2NCwKICAgICAgICB9LAogICAgICB9OwoKICAgICAgbGV0IHBlcmNlbnREb25lID0gZmlsZURhdGEuYnl0ZUxlbmd0aCA9PT0gMCA/CiAgICAgICAgICAxMDAgOgogICAgICAgICAgTWF0aC5yb3VuZCgocG9zaXRpb24gLyBmaWxlRGF0YS5ieXRlTGVuZ3RoKSAqIDEwMCk7CiAgICAgIHBlcmNlbnQudGV4dENvbnRlbnQgPSBgJHtwZXJjZW50RG9uZX0lIGRvbmVgOwoKICAgIH0gd2hpbGUgKHBvc2l0aW9uIDwgZmlsZURhdGEuYnl0ZUxlbmd0aCk7CiAgfQoKICAvLyBBbGwgZG9uZS4KICB5aWVsZCB7CiAgICByZXNwb25zZTogewogICAgICBhY3Rpb246ICdjb21wbGV0ZScsCiAgICB9CiAgfTsKfQoKc2NvcGUuZ29vZ2xlID0gc2NvcGUuZ29vZ2xlIHx8IHt9OwpzY29wZS5nb29nbGUuY29sYWIgPSBzY29wZS5nb29nbGUuY29sYWIgfHwge307CnNjb3BlLmdvb2dsZS5jb2xhYi5fZmlsZXMgPSB7CiAgX3VwbG9hZEZpbGVzLAogIF91cGxvYWRGaWxlc0NvbnRpbnVlLAp9Owp9KShzZWxmKTsK",
              "ok": true,
              "headers": [
                [
                  "content-type",
                  "application/javascript"
                ]
              ],
              "status": 200,
              "status_text": "OK"
            }
          },
          "height": 94
        },
        "outputId": "5ae9216e-d6f5-4574-d640-d6b073a01500"
      },
      "source": [
        "os.chdir('/content')\n",
        "do_upload = True\n",
        "try:\n",
        "  os.mkdir('work_dir')\n",
        "except:\n",
        "  do_upload=False\n",
        "  print(\"Directoria ja existe!\")\n",
        "\n",
        "  # fazer o upload manual da trajectoria\n",
        "\n",
        "os.chdir('/content/work_dir')\n",
        "\n",
        "# regressa a work_dir\n",
        "os.chdir('/content/work_dir')\n",
        "\n",
        "if do_upload:\n",
        "  from google.colab import files\n",
        "  uploaded = files.upload()\n",
        "  for fn in uploaded.keys():\n",
        "    print('User uploaded file \"{name}\" with length {length} bytes'.format(\n",
        "          name=fn, length=len(uploaded[fn])))\n",
        "    trj_fn = f'/content/work_dir/{fn}'\n",
        "\n",
        "delta_t_trj = 0.05 #janela de tempo entre pontos da trjactoria (ver ficheiro de input) em ps\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<IPython.core.display.HTML object>"
            ],
            "text/html": [
              "\n",
              "     <input type=\"file\" id=\"files-15539fcb-4802-4a9d-bd7e-7f79d826a395\" name=\"files[]\" multiple disabled\n",
              "        style=\"border:none\" />\n",
              "     <output id=\"result-15539fcb-4802-4a9d-bd7e-7f79d826a395\">\n",
              "      Upload widget is only available when the cell has been executed in the\n",
              "      current browser session. Please rerun this cell to enable.\n",
              "      </output>\n",
              "      <script src=\"/nbextensions/google.colab/files.js\"></script> "
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Saving sys03-mtd03.xtb.trj to sys03-mtd03.xtb.trj\n",
            "User uploaded file \"sys03-mtd03.xtb.trj\" with length 94540000 bytes\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Outline do procedimento\n",
        "\n",
        "* Extrair da trajectória as geometrias da frame _n_ (estado final, produtos) e _n-1_ (estado inicial, reagentes).\n",
        "* Isolar as moléculas envolvidas na transformação.\n",
        "* Correr o LLM entre os estados inicial e final\n",
        "  * Garantir que não há descontinuidades ao longo do LLM\n",
        "* identificar dois pontos próximos do máximo de energia ao longo do LLM: P1 e P2.\n",
        "* correr o `xtb` com um calculo do tipo `--path` para encontrar uma geometria aceitavel para o estado de transição.\n",
        "  * Verificar que o calculo convergiu: analisar o path final\n",
        "* Correr o `xtb`com um cálculo do tipo `--hess` para a geometria do estado de transição obtida no ponto anterior\n",
        "  * Garantir que o estado de transição tem uma e apenas uma frequência imaginária relevante.\n",
        "* Correr o `xtb`com um cálculo do tipo `--ohess` para as geometruas P1 (reagente) e P2 (produto)\n",
        "  * Verifica que cada optimização convergiu para lados diferentes do caminho reaccional, caso uma das optimizações falhe este objectivo, escolha um ponto ligeiramente mais afastado do TS.\n",
        "* Correr a função `compile_reaction_data` fornecendo os outputs do reagente, estado de transição e produto.\n"
      ],
      "metadata": {
        "id": "w9m-oNNxiJkm"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Reacçao 1\n",
        "\n",
        "* Frames 37 -> 38\n",
        "* Reacção 2CO -> OCCO\n",
        "* Atomos envolvidos: 123 e 125"
      ],
      "metadata": {
        "id": "q6c1d4xwh1Bx"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# extrair os estados inicial e final\n",
        "p1_fn = 'reac01-est01-guess.xyz'\n",
        "p2_fn = 'reac01-est02-guess.xyz'\n",
        "\n",
        "\n",
        "extract_geo('/content/work_dir/sys03-mtd03.xtb.trj',36,p1_fn) # n+1\n",
        "extract_geo('/content/work_dir/sys03-mtd03.xtb.trj',39,p2_fn) # n+1\n",
        "\n",
        "m1 = Molecule(p1_fn)\n",
        "m2 = Molecule(p2_fn)\n",
        "\n",
        "m1.show()\n",
        "m2.show()"
      ],
      "metadata": {
        "id": "JC3akGebu9rB",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "870406e5-077a-4d23-dbde-a5c45badcddb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581882493002732\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581882493002732\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581882493002732 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581882493002732\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581882493002732 = $3Dmol.createViewer($(\"#3dmolviewer_16581882493002732\"),{backgroundColor:\"white\"});\nviewer_16581882493002732.zoomTo();\n\tviewer_16581882493002732.addModel(\"140\\n Created by Molecule Class\\nO           3.216976        -2.777220        -2.759669\\nH           4.091214        -2.761867        -2.315667\\nH           2.628334        -2.151939        -2.376342\\nO           4.285721         0.283123        -2.350455\\nH           4.865572         0.253856        -1.562204\\nH           4.072207         1.209004        -2.514543\\nO           0.898990        -0.225036        -1.806679\\nH           0.057549         0.076628        -1.610965\\nH           0.942879        -1.052568        -1.270895\\nO           4.920381        -2.597553         1.097754\\nH           4.225101        -2.010416         1.500753\\nH           5.437470        -2.077579         0.498244\\nO           3.252676        -0.814617         2.476636\\nH           2.480254        -1.300460         2.723499\\nH           3.024902        -0.018193         1.951080\\nO           2.560390        -2.460081        -8.126636\\nH           3.194414        -1.909391        -7.702001\\nH           1.663994        -2.143295        -7.954422\\nO           2.720013        -3.520475         2.994349\\nH           2.623569        -4.473932         3.050765\\nH           3.550322        -3.161086         2.852462\\nO           2.178872        -1.161489        -4.893285\\nH           2.640320        -1.607141        -4.160882\\nH           1.391298        -0.704662        -4.546313\\nO          -5.451170        -4.067678        -4.543267\\nH          -5.099603        -3.158243        -4.506453\\nH          -6.357724        -4.035279        -4.923945\\nO          -0.315057         4.752560        -4.589842\\nH          -1.177266         4.465439        -4.732542\\nH           0.214640         5.016173        -5.356997\\nN           0.271867        -2.293425         1.375661\\nH          -0.291017        -3.148885         1.255097\\nH           0.838545        -2.263317         0.654929\\nH           0.906594        -2.449078         2.172295\\nN          -0.625484         4.727212         5.262823\\nH          -0.165663         4.061686         5.829875\\nH          -1.598157         4.880024         5.517009\\nH          -0.356334         4.568042         4.309931\\nN          -3.683124        -5.764002         3.842019\\nH          -4.156016        -4.971357         3.408158\\nH          -4.233273        -6.616864         4.042914\\nH          -3.390912        -5.297572         4.710313\\nN          -2.136149        -2.920234        -6.319455\\nH          -1.366263        -2.490942        -5.887003\\nH          -1.711675        -3.090336        -7.312395\\nH          -2.962026        -2.386347        -6.477343\\nN          -0.907702         6.557596         0.288163\\nH          -1.570968         7.088454        -0.308470\\nH          -0.386634         7.062976         0.975668\\nH          -0.257073         6.119508        -0.258479\\nN          -1.592985         2.610620        -0.575432\\nH          -0.761033         2.104411        -0.325710\\nH          -1.588500         3.559686        -0.153121\\nH          -2.405060         2.148519        -0.208137\\nN          -2.333234         0.528804         1.931287\\nH          -3.318677         0.383944         2.198472\\nH          -1.791049        -0.313351         1.517175\\nH          -1.818274         0.771404         2.835169\\nN          -1.271565         1.293734        -6.596696\\nH          -1.865835         1.056036        -5.802257\\nH          -1.495194         0.668538        -7.334172\\nH          -0.334720         1.093715        -6.300789\\nN           0.403720        -5.796013         2.137949\\nH          -0.476945        -5.468064         2.398612\\nH           0.908420        -5.271070         1.451066\\nH           0.313157        -6.790449         2.003848\\nN           4.461615         6.003720         2.178722\\nH           4.373586         6.659133         1.440852\\nH           4.118655         6.550078         3.005148\\nH           5.454359         5.778556         2.257689\\nC          -4.507605        -0.148268        -6.774667\\nH          -5.037811        -0.397111        -7.682999\\nH          -3.971110         0.752141        -6.976833\\nH          -5.368363        -0.111059        -5.943045\\nH          -3.855437        -0.970022        -6.373035\\nC           2.096329         2.870502        -3.043705\\nH           2.891404         2.781881        -3.752930\\nH           2.251208         2.003211        -2.404481\\nH           2.251119         3.749003        -2.523138\\nH           1.124582         2.921933        -3.456686\\nC          -4.356810         5.569206         0.835607\\nH          -4.643999         5.437048        -0.211856\\nH          -3.334772         5.792970         0.713510\\nH          -4.916173         6.374354         1.320020\\nH          -4.491234         4.583265         1.411037\\nC           2.148799         2.158989         3.622469\\nH           2.118262         2.463158         2.573796\\nH           2.264696         1.120638         3.721604\\nH           2.946155         2.610664         4.106693\\nH           1.247874         2.346746         4.126167\\nC           0.913653        -4.656535        -2.410697\\nH           0.413309        -3.805075        -1.767437\\nH           0.215763        -5.477685        -2.365125\\nH           1.189363        -4.286276        -3.403351\\nH           1.868601        -4.857168        -1.778104\\nC           1.043452        -4.699141        -7.225911\\nH           1.960105        -5.038654        -7.679174\\nH           0.696162        -3.733209        -7.627025\\nH           0.281050        -5.442794        -7.405839\\nH           1.073362        -4.566014        -6.179202\\nC          -6.044158         1.085878         3.862437\\nH          -6.852472         0.628334         4.407437\\nH          -6.348543         1.973666         3.464904\\nH          -5.153395         1.196590         4.463486\\nH          -5.805553         0.334952         3.083096\\nC           6.221664        -5.034634        -2.836603\\nH           6.713851        -4.538254        -2.065973\\nH           5.371506        -5.631813        -2.466469\\nH           5.940785        -4.437848        -3.746521\\nH           7.022483        -5.743836        -3.063532\\nC           7.176452         1.737302         4.607454\\nH           7.864513         1.363657         5.369917\\nH           6.589291         2.647815         4.872767\\nH           7.796630         2.086561         3.837943\\nH           6.629133         0.869071         4.272022\\nC          -4.169410         6.457015         4.107677\\nH          -4.054766         6.701181         5.173293\\nH          -4.198660         5.449981         3.811027\\nH          -5.154465         6.842962         3.857549\\nH          -3.371285         6.960913         3.601901\\nC          -5.862810        -3.282703         1.349667\\nO          -6.310672        -4.301746         1.538045\\nC           0.489280        -5.853018         6.679023\\nO           0.686817        -6.353458         5.687463\\nC           0.966134        -3.256543         4.942166\\nO           0.217765        -2.479663         4.560904\\nC           6.024858         2.612817        -0.365746\\nO           5.617882         3.531069        -0.880409\\nC          -2.137894        -2.846539        -0.092927\\nO          -3.237601        -3.076126         0.159953\\nH          -3.323670         2.810134         7.527616\\nH          -3.906745         3.183881         7.760415\\nH           2.320479         2.149529        -8.661968\\nH           2.220930         1.472242        -8.981095\\nH           5.322185         6.643593        -3.981932\\nH           5.131346         7.225166        -3.456752\\nH           5.010185        -6.208038         0.576043\\nH           4.776658        -6.752966         1.159790\\nH          -6.722915         4.231131         2.771251\\nH          -6.909911         4.242799         3.515577\\n\");\n\tviewer_16581882493002732.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581882493002732.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581882493002732.zoomTo();\nviewer_16581882493002732.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581882493002732\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581882493002732\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581882493002732 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581882493002732\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581882493002732 = $3Dmol.createViewer($(\"#3dmolviewer_16581882493002732\"),{backgroundColor:\"white\"});\n",
              "viewer_16581882493002732.zoomTo();\n",
              "\tviewer_16581882493002732.addModel(\"140\\n Created by Molecule Class\\nO           3.216976        -2.777220        -2.759669\\nH           4.091214        -2.761867        -2.315667\\nH           2.628334        -2.151939        -2.376342\\nO           4.285721         0.283123        -2.350455\\nH           4.865572         0.253856        -1.562204\\nH           4.072207         1.209004        -2.514543\\nO           0.898990        -0.225036        -1.806679\\nH           0.057549         0.076628        -1.610965\\nH           0.942879        -1.052568        -1.270895\\nO           4.920381        -2.597553         1.097754\\nH           4.225101        -2.010416         1.500753\\nH           5.437470        -2.077579         0.498244\\nO           3.252676        -0.814617         2.476636\\nH           2.480254        -1.300460         2.723499\\nH           3.024902        -0.018193         1.951080\\nO           2.560390        -2.460081        -8.126636\\nH           3.194414        -1.909391        -7.702001\\nH           1.663994        -2.143295        -7.954422\\nO           2.720013        -3.520475         2.994349\\nH           2.623569        -4.473932         3.050765\\nH           3.550322        -3.161086         2.852462\\nO           2.178872        -1.161489        -4.893285\\nH           2.640320        -1.607141        -4.160882\\nH           1.391298        -0.704662        -4.546313\\nO          -5.451170        -4.067678        -4.543267\\nH          -5.099603        -3.158243        -4.506453\\nH          -6.357724        -4.035279        -4.923945\\nO          -0.315057         4.752560        -4.589842\\nH          -1.177266         4.465439        -4.732542\\nH           0.214640         5.016173        -5.356997\\nN           0.271867        -2.293425         1.375661\\nH          -0.291017        -3.148885         1.255097\\nH           0.838545        -2.263317         0.654929\\nH           0.906594        -2.449078         2.172295\\nN          -0.625484         4.727212         5.262823\\nH          -0.165663         4.061686         5.829875\\nH          -1.598157         4.880024         5.517009\\nH          -0.356334         4.568042         4.309931\\nN          -3.683124        -5.764002         3.842019\\nH          -4.156016        -4.971357         3.408158\\nH          -4.233273        -6.616864         4.042914\\nH          -3.390912        -5.297572         4.710313\\nN          -2.136149        -2.920234        -6.319455\\nH          -1.366263        -2.490942        -5.887003\\nH          -1.711675        -3.090336        -7.312395\\nH          -2.962026        -2.386347        -6.477343\\nN          -0.907702         6.557596         0.288163\\nH          -1.570968         7.088454        -0.308470\\nH          -0.386634         7.062976         0.975668\\nH          -0.257073         6.119508        -0.258479\\nN          -1.592985         2.610620        -0.575432\\nH          -0.761033         2.104411        -0.325710\\nH          -1.588500         3.559686        -0.153121\\nH          -2.405060         2.148519        -0.208137\\nN          -2.333234         0.528804         1.931287\\nH          -3.318677         0.383944         2.198472\\nH          -1.791049        -0.313351         1.517175\\nH          -1.818274         0.771404         2.835169\\nN          -1.271565         1.293734        -6.596696\\nH          -1.865835         1.056036        -5.802257\\nH          -1.495194         0.668538        -7.334172\\nH          -0.334720         1.093715        -6.300789\\nN           0.403720        -5.796013         2.137949\\nH          -0.476945        -5.468064         2.398612\\nH           0.908420        -5.271070         1.451066\\nH           0.313157        -6.790449         2.003848\\nN           4.461615         6.003720         2.178722\\nH           4.373586         6.659133         1.440852\\nH           4.118655         6.550078         3.005148\\nH           5.454359         5.778556         2.257689\\nC          -4.507605        -0.148268        -6.774667\\nH          -5.037811        -0.397111        -7.682999\\nH          -3.971110         0.752141        -6.976833\\nH          -5.368363        -0.111059        -5.943045\\nH          -3.855437        -0.970022        -6.373035\\nC           2.096329         2.870502        -3.043705\\nH           2.891404         2.781881        -3.752930\\nH           2.251208         2.003211        -2.404481\\nH           2.251119         3.749003        -2.523138\\nH           1.124582         2.921933        -3.456686\\nC          -4.356810         5.569206         0.835607\\nH          -4.643999         5.437048        -0.211856\\nH          -3.334772         5.792970         0.713510\\nH          -4.916173         6.374354         1.320020\\nH          -4.491234         4.583265         1.411037\\nC           2.148799         2.158989         3.622469\\nH           2.118262         2.463158         2.573796\\nH           2.264696         1.120638         3.721604\\nH           2.946155         2.610664         4.106693\\nH           1.247874         2.346746         4.126167\\nC           0.913653        -4.656535        -2.410697\\nH           0.413309        -3.805075        -1.767437\\nH           0.215763        -5.477685        -2.365125\\nH           1.189363        -4.286276        -3.403351\\nH           1.868601        -4.857168        -1.778104\\nC           1.043452        -4.699141        -7.225911\\nH           1.960105        -5.038654        -7.679174\\nH           0.696162        -3.733209        -7.627025\\nH           0.281050        -5.442794        -7.405839\\nH           1.073362        -4.566014        -6.179202\\nC          -6.044158         1.085878         3.862437\\nH          -6.852472         0.628334         4.407437\\nH          -6.348543         1.973666         3.464904\\nH          -5.153395         1.196590         4.463486\\nH          -5.805553         0.334952         3.083096\\nC           6.221664        -5.034634        -2.836603\\nH           6.713851        -4.538254        -2.065973\\nH           5.371506        -5.631813        -2.466469\\nH           5.940785        -4.437848        -3.746521\\nH           7.022483        -5.743836        -3.063532\\nC           7.176452         1.737302         4.607454\\nH           7.864513         1.363657         5.369917\\nH           6.589291         2.647815         4.872767\\nH           7.796630         2.086561         3.837943\\nH           6.629133         0.869071         4.272022\\nC          -4.169410         6.457015         4.107677\\nH          -4.054766         6.701181         5.173293\\nH          -4.198660         5.449981         3.811027\\nH          -5.154465         6.842962         3.857549\\nH          -3.371285         6.960913         3.601901\\nC          -5.862810        -3.282703         1.349667\\nO          -6.310672        -4.301746         1.538045\\nC           0.489280        -5.853018         6.679023\\nO           0.686817        -6.353458         5.687463\\nC           0.966134        -3.256543         4.942166\\nO           0.217765        -2.479663         4.560904\\nC           6.024858         2.612817        -0.365746\\nO           5.617882         3.531069        -0.880409\\nC          -2.137894        -2.846539        -0.092927\\nO          -3.237601        -3.076126         0.159953\\nH          -3.323670         2.810134         7.527616\\nH          -3.906745         3.183881         7.760415\\nH           2.320479         2.149529        -8.661968\\nH           2.220930         1.472242        -8.981095\\nH           5.322185         6.643593        -3.981932\\nH           5.131346         7.225166        -3.456752\\nH           5.010185        -6.208038         0.576043\\nH           4.776658        -6.752966         1.159790\\nH          -6.722915         4.231131         2.771251\\nH          -6.909911         4.242799         3.515577\\n\");\n",
              "\tviewer_16581882493002732.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581882493002732.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581882493002732.zoomTo();\n",
              "viewer_16581882493002732.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581882493037007\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581882493037007\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581882493037007 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581882493037007\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581882493037007 = $3Dmol.createViewer($(\"#3dmolviewer_16581882493037007\"),{backgroundColor:\"white\"});\nviewer_16581882493037007.zoomTo();\n\tviewer_16581882493037007.addModel(\"140\\n Created by Molecule Class\\nO           3.484613        -2.556138        -4.017458\\nH           4.002742        -1.971442        -3.457168\\nH           2.798247        -2.939700        -3.393628\\nO           3.990607        -0.267023        -1.803510\\nH           4.549250        -0.028901        -1.068555\\nH           3.710720         0.481643        -2.380772\\nO           0.573188        -0.203762        -1.022442\\nH          -0.188499        -0.501082        -0.521664\\nH           0.495510         0.779343        -0.714436\\nO           4.325098        -2.534624         0.400153\\nH           3.831785        -1.714878         0.569479\\nH           4.579032        -2.775534         1.305476\\nO           2.900256        -1.010810         2.413203\\nH           3.410733        -1.335647         3.152930\\nH           2.345731        -1.751226         2.117519\\nO           3.561755        -2.863945        -8.113452\\nH           2.737949        -3.361342        -8.327616\\nH           3.542890        -1.968585        -7.648364\\nO           3.150445        -3.469164         3.706493\\nH           2.547771        -3.579996         2.938565\\nH           3.096242        -4.197936         4.288614\\nO           2.523048        -0.696654        -5.884399\\nH           2.921782        -1.468573        -5.447038\\nH           2.206668        -0.153758        -5.148921\\nO          -4.698023        -3.535119        -4.348979\\nH          -3.794352        -3.096193        -4.260914\\nH          -4.962034        -3.832551        -5.225852\\nO          -1.285422         3.502884        -4.537482\\nH          -1.703549         4.400426        -4.394338\\nH          -1.634392         2.909777        -3.853437\\nN           1.534170        -2.569164         0.836531\\nH           1.514412        -1.838893         0.224696\\nH           2.269134        -3.273676         0.537291\\nH           0.753173        -3.097838         0.769452\\nN          -0.825807         3.910093         5.256543\\nH          -1.811367         4.163512         5.383438\\nH          -0.307540         4.592207         5.821096\\nH          -0.641878         3.990556         4.285642\\nN          -3.448136        -5.603241         4.595674\\nH          -3.720313        -6.584742         4.938568\\nH          -3.887176        -5.275380         3.823984\\nH          -2.403188        -5.684223         4.465252\\nN          -1.911734        -2.257190        -7.752963\\nH          -1.156832        -2.785389        -7.327590\\nH          -1.593114        -1.240877        -7.647716\\nH          -2.769510        -2.276211        -7.223778\\nN          -1.138229         7.110255        -0.338033\\nH          -1.838955         7.052830        -1.097622\\nH          -0.200096         7.323766        -0.656870\\nH          -1.180799         6.368931         0.281304\\nN          -1.065016         3.429147        -0.466603\\nH          -1.158585         2.743770        -1.216658\\nH          -2.013678         3.860257        -0.302463\\nH          -0.406565         4.075391        -0.758041\\nN          -0.903775         0.179125         1.166516\\nH          -1.365161        -0.744920         0.901542\\nH          -0.073628         0.021740         1.709916\\nH          -1.473419         0.695434         1.941401\\nN          -1.550240         1.265033        -7.098151\\nH          -1.137234         0.690403        -7.849983\\nH          -1.477131         2.189102        -7.399510\\nH          -1.079894         1.262203        -6.227987\\nN           0.314004        -6.046990         1.519975\\nH           0.698659        -5.741950         2.377260\\nH          -0.597689        -6.276406         1.400475\\nH           0.925738        -6.806653         1.179738\\nN           4.099439         6.593167         2.567707\\nH           4.132056         6.966175         3.530383\\nH           4.812025         7.140666         2.005229\\nH           3.185846         6.857125         2.240974\\nC          -4.953487         0.192017        -6.525480\\nH          -4.024973        -0.208533        -6.336326\\nH          -4.959291         1.238236        -6.625404\\nH          -5.559031        -0.260672        -7.378224\\nH          -5.506911         0.007030        -5.541576\\nC           2.414788         3.373086        -2.956369\\nH           2.121314         2.762184        -3.837744\\nH           1.963675         2.747628        -2.155718\\nH           3.486506         3.410818        -2.835111\\nH           1.885673         4.263224        -3.087049\\nC          -4.388928         4.558357         0.510524\\nH          -5.402962         4.345195         0.181119\\nH          -4.439623         4.854650         1.562245\\nH          -3.670314         3.822479         0.303499\\nH          -4.089719         5.444123        -0.121337\\nC           2.808891         2.647651         2.694061\\nH           2.978932         1.843545         2.015632\\nH           2.644832         3.526196         2.095257\\nH           1.916399         2.437556         3.246036\\nH           3.567332         2.881709         3.379508\\nC           0.886582        -4.353161        -2.818580\\nH           0.363350        -3.534735        -3.400496\\nH           0.277939        -4.710893        -1.988666\\nH           1.083624        -5.176241        -3.433122\\nH           1.888050        -4.005597        -2.388104\\nC           0.406777        -5.089173        -6.096637\\nH          -0.585271        -5.081418        -5.564949\\nH           0.959119        -4.238210        -5.661678\\nH           0.857794        -6.090915        -6.024824\\nH           0.154299        -4.819070        -7.116855\\nC          -5.448507         1.121093         4.040816\\nH          -5.964812         0.291532         4.512817\\nH          -4.657362         0.648911         3.606876\\nH          -6.203557         1.497263         3.291693\\nH          -5.217467         1.858142         4.731766\\nC           6.022475        -4.536813        -2.537454\\nH           6.246396        -4.755629        -1.551832\\nH           5.457806        -3.633172        -2.387834\\nH           6.927427        -4.298675        -3.089431\\nH           5.406007        -5.282877        -3.019281\\nC           6.549052         1.499819         4.745451\\nH           5.963677         1.457345         3.809526\\nH           5.947123         0.781308         5.359825\\nH           7.491167         0.905931         4.631715\\nH           6.788768         2.515397         5.127509\\nC          -3.639696         5.980779         4.458195\\nH          -2.950174         6.582414         3.849417\\nH          -3.883317         6.663893         5.242677\\nH          -4.443377         5.658671         3.803197\\nH          -3.142977         5.159287         4.892022\\nC          -6.137478        -2.992705         2.837022\\nO          -7.186632        -3.153239         2.454350\\nC           0.122249        -5.353459         6.864350\\nO          -0.672465        -6.216738         6.882098\\nC           0.104168        -4.210207         5.607139\\nO           0.383115        -4.225366         4.482529\\nC           6.026104         2.831845        -0.905720\\nO           5.321107         3.686094        -0.627109\\nC          -2.417103        -2.282295        -0.035836\\nO          -2.663099        -3.300319         0.432939\\nH          -4.387562         4.108417         6.173012\\nH          -4.274215         3.477086         6.765479\\nH           1.627089         1.664121        -8.119839\\nH           1.928917         2.434464        -8.097854\\nH           3.980529         7.621033        -2.041316\\nH           4.320699         7.157785        -1.574567\\nH           5.153954        -7.909970         2.121692\\nH           5.365688        -7.527040         1.521990\\nH          -6.915028         3.856551         2.293194\\nH          -7.573760         4.279361         2.440703\\n\");\n\tviewer_16581882493037007.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581882493037007.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581882493037007.zoomTo();\nviewer_16581882493037007.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581882493037007\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581882493037007\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581882493037007 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581882493037007\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581882493037007 = $3Dmol.createViewer($(\"#3dmolviewer_16581882493037007\"),{backgroundColor:\"white\"});\n",
              "viewer_16581882493037007.zoomTo();\n",
              "\tviewer_16581882493037007.addModel(\"140\\n Created by Molecule Class\\nO           3.484613        -2.556138        -4.017458\\nH           4.002742        -1.971442        -3.457168\\nH           2.798247        -2.939700        -3.393628\\nO           3.990607        -0.267023        -1.803510\\nH           4.549250        -0.028901        -1.068555\\nH           3.710720         0.481643        -2.380772\\nO           0.573188        -0.203762        -1.022442\\nH          -0.188499        -0.501082        -0.521664\\nH           0.495510         0.779343        -0.714436\\nO           4.325098        -2.534624         0.400153\\nH           3.831785        -1.714878         0.569479\\nH           4.579032        -2.775534         1.305476\\nO           2.900256        -1.010810         2.413203\\nH           3.410733        -1.335647         3.152930\\nH           2.345731        -1.751226         2.117519\\nO           3.561755        -2.863945        -8.113452\\nH           2.737949        -3.361342        -8.327616\\nH           3.542890        -1.968585        -7.648364\\nO           3.150445        -3.469164         3.706493\\nH           2.547771        -3.579996         2.938565\\nH           3.096242        -4.197936         4.288614\\nO           2.523048        -0.696654        -5.884399\\nH           2.921782        -1.468573        -5.447038\\nH           2.206668        -0.153758        -5.148921\\nO          -4.698023        -3.535119        -4.348979\\nH          -3.794352        -3.096193        -4.260914\\nH          -4.962034        -3.832551        -5.225852\\nO          -1.285422         3.502884        -4.537482\\nH          -1.703549         4.400426        -4.394338\\nH          -1.634392         2.909777        -3.853437\\nN           1.534170        -2.569164         0.836531\\nH           1.514412        -1.838893         0.224696\\nH           2.269134        -3.273676         0.537291\\nH           0.753173        -3.097838         0.769452\\nN          -0.825807         3.910093         5.256543\\nH          -1.811367         4.163512         5.383438\\nH          -0.307540         4.592207         5.821096\\nH          -0.641878         3.990556         4.285642\\nN          -3.448136        -5.603241         4.595674\\nH          -3.720313        -6.584742         4.938568\\nH          -3.887176        -5.275380         3.823984\\nH          -2.403188        -5.684223         4.465252\\nN          -1.911734        -2.257190        -7.752963\\nH          -1.156832        -2.785389        -7.327590\\nH          -1.593114        -1.240877        -7.647716\\nH          -2.769510        -2.276211        -7.223778\\nN          -1.138229         7.110255        -0.338033\\nH          -1.838955         7.052830        -1.097622\\nH          -0.200096         7.323766        -0.656870\\nH          -1.180799         6.368931         0.281304\\nN          -1.065016         3.429147        -0.466603\\nH          -1.158585         2.743770        -1.216658\\nH          -2.013678         3.860257        -0.302463\\nH          -0.406565         4.075391        -0.758041\\nN          -0.903775         0.179125         1.166516\\nH          -1.365161        -0.744920         0.901542\\nH          -0.073628         0.021740         1.709916\\nH          -1.473419         0.695434         1.941401\\nN          -1.550240         1.265033        -7.098151\\nH          -1.137234         0.690403        -7.849983\\nH          -1.477131         2.189102        -7.399510\\nH          -1.079894         1.262203        -6.227987\\nN           0.314004        -6.046990         1.519975\\nH           0.698659        -5.741950         2.377260\\nH          -0.597689        -6.276406         1.400475\\nH           0.925738        -6.806653         1.179738\\nN           4.099439         6.593167         2.567707\\nH           4.132056         6.966175         3.530383\\nH           4.812025         7.140666         2.005229\\nH           3.185846         6.857125         2.240974\\nC          -4.953487         0.192017        -6.525480\\nH          -4.024973        -0.208533        -6.336326\\nH          -4.959291         1.238236        -6.625404\\nH          -5.559031        -0.260672        -7.378224\\nH          -5.506911         0.007030        -5.541576\\nC           2.414788         3.373086        -2.956369\\nH           2.121314         2.762184        -3.837744\\nH           1.963675         2.747628        -2.155718\\nH           3.486506         3.410818        -2.835111\\nH           1.885673         4.263224        -3.087049\\nC          -4.388928         4.558357         0.510524\\nH          -5.402962         4.345195         0.181119\\nH          -4.439623         4.854650         1.562245\\nH          -3.670314         3.822479         0.303499\\nH          -4.089719         5.444123        -0.121337\\nC           2.808891         2.647651         2.694061\\nH           2.978932         1.843545         2.015632\\nH           2.644832         3.526196         2.095257\\nH           1.916399         2.437556         3.246036\\nH           3.567332         2.881709         3.379508\\nC           0.886582        -4.353161        -2.818580\\nH           0.363350        -3.534735        -3.400496\\nH           0.277939        -4.710893        -1.988666\\nH           1.083624        -5.176241        -3.433122\\nH           1.888050        -4.005597        -2.388104\\nC           0.406777        -5.089173        -6.096637\\nH          -0.585271        -5.081418        -5.564949\\nH           0.959119        -4.238210        -5.661678\\nH           0.857794        -6.090915        -6.024824\\nH           0.154299        -4.819070        -7.116855\\nC          -5.448507         1.121093         4.040816\\nH          -5.964812         0.291532         4.512817\\nH          -4.657362         0.648911         3.606876\\nH          -6.203557         1.497263         3.291693\\nH          -5.217467         1.858142         4.731766\\nC           6.022475        -4.536813        -2.537454\\nH           6.246396        -4.755629        -1.551832\\nH           5.457806        -3.633172        -2.387834\\nH           6.927427        -4.298675        -3.089431\\nH           5.406007        -5.282877        -3.019281\\nC           6.549052         1.499819         4.745451\\nH           5.963677         1.457345         3.809526\\nH           5.947123         0.781308         5.359825\\nH           7.491167         0.905931         4.631715\\nH           6.788768         2.515397         5.127509\\nC          -3.639696         5.980779         4.458195\\nH          -2.950174         6.582414         3.849417\\nH          -3.883317         6.663893         5.242677\\nH          -4.443377         5.658671         3.803197\\nH          -3.142977         5.159287         4.892022\\nC          -6.137478        -2.992705         2.837022\\nO          -7.186632        -3.153239         2.454350\\nC           0.122249        -5.353459         6.864350\\nO          -0.672465        -6.216738         6.882098\\nC           0.104168        -4.210207         5.607139\\nO           0.383115        -4.225366         4.482529\\nC           6.026104         2.831845        -0.905720\\nO           5.321107         3.686094        -0.627109\\nC          -2.417103        -2.282295        -0.035836\\nO          -2.663099        -3.300319         0.432939\\nH          -4.387562         4.108417         6.173012\\nH          -4.274215         3.477086         6.765479\\nH           1.627089         1.664121        -8.119839\\nH           1.928917         2.434464        -8.097854\\nH           3.980529         7.621033        -2.041316\\nH           4.320699         7.157785        -1.574567\\nH           5.153954        -7.909970         2.121692\\nH           5.365688        -7.527040         1.521990\\nH          -6.915028         3.856551         2.293194\\nH          -7.573760         4.279361         2.440703\\n\");\n",
              "\tviewer_16581882493037007.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581882493037007.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581882493037007.zoomTo();\n",
              "viewer_16581882493037007.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# NOVIDADE: Extraimos apenas os átomos de interesse, i.e. aqueles que estão nas moléculas dos reagentes ou dos produtos\n",
        "# NB: o python começa acontar em 0, pelo que temos que subtrair 1 a cada um dos indíces dos átomos de interese\n",
        "#     podemos fazer isto no próprio comando, como abaixo\n",
        "isolate_molecules_with_atoms(p1_fn,[123-1,125-1],'reac1-state1_simple.xyz')\n",
        "isolate_molecules_with_atoms(p2_fn,[123-1,125-1],'reac1-state2_simple.xyz')"
      ],
      "metadata": {
        "id": "QEfQjafcDuSF"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "m1 = Molecule('reac1-state1_simple.xyz')\n",
        "m2 = Molecule('reac1-state2_simple.xyz')\n",
        "\n",
        "m1.show()\n",
        "m2.show()"
      ],
      "metadata": {
        "id": "OpF3sVOyEKi0",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "00198152-dad9-4b79-a2ce-f6c1456f6931"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581882550563784\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581882550563784\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581882550563784 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581882550563784\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581882550563784 = $3Dmol.createViewer($(\"#3dmolviewer_16581882550563784\"),{backgroundColor:\"white\"});\nviewer_16581882550563784.zoomTo();\n\tviewer_16581882550563784.addModel(\"4\\n Created by Molecule Class\\nC          -0.100719        -1.367347         1.211634\\nO           0.096818        -1.867787         0.220074\\nC           0.376135         1.229127        -0.525223\\nO          -0.372234         2.006007        -0.906485\\n\");\n\tviewer_16581882550563784.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581882550563784.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581882550563784.zoomTo();\nviewer_16581882550563784.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581882550563784\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581882550563784\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581882550563784 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581882550563784\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581882550563784 = $3Dmol.createViewer($(\"#3dmolviewer_16581882550563784\"),{backgroundColor:\"white\"});\n",
              "viewer_16581882550563784.zoomTo();\n",
              "\tviewer_16581882550563784.addModel(\"4\\n Created by Molecule Class\\nC          -0.100719        -1.367347         1.211634\\nO           0.096818        -1.867787         0.220074\\nC           0.376135         1.229127        -0.525223\\nO          -0.372234         2.006007        -0.906485\\n\");\n",
              "\tviewer_16581882550563784.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581882550563784.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581882550563784.zoomTo();\n",
              "viewer_16581882550563784.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581882550597527\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581882550597527\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581882550597527 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581882550597527\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581882550597527 = $3Dmol.createViewer($(\"#3dmolviewer_16581882550597527\"),{backgroundColor:\"white\"});\nviewer_16581882550597527.zoomTo();\n\tviewer_16581882550597527.addModel(\"4\\n Created by Molecule Class\\nC           0.137982        -0.352016         0.905321\\nO          -0.656731        -1.215296         0.923069\\nC           0.119902         0.791235        -0.351890\\nO           0.398848         0.776077        -1.476500\\n\");\n\tviewer_16581882550597527.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581882550597527.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581882550597527.zoomTo();\nviewer_16581882550597527.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581882550597527\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581882550597527\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581882550597527 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581882550597527\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581882550597527 = $3Dmol.createViewer($(\"#3dmolviewer_16581882550597527\"),{backgroundColor:\"white\"});\n",
              "viewer_16581882550597527.zoomTo();\n",
              "\tviewer_16581882550597527.addModel(\"4\\n Created by Molecule Class\\nC           0.137982        -0.352016         0.905321\\nO          -0.656731        -1.215296         0.923069\\nC           0.119902         0.791235        -0.351890\\nO           0.398848         0.776077        -1.476500\\n\");\n",
              "\tviewer_16581882550597527.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581882550597527.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581882550597527.zoomTo();\n",
              "viewer_16581882550597527.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "llm('reac1-state1_simple.xyz', 'reac1-state2_simple.xyz', 'reac01-lmm-teste', n=30)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 817
        },
        "id": "Bz_oJ8mCpBo6",
        "outputId": "e265da64-a2db-490e-e496-91b2d6259636"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "    0        -12.24380\n",
            "    1        -12.24172\n",
            "    2        -12.22990\n",
            "    3        -12.20786\n",
            "    4        -12.17556\n",
            "    5        -12.13352\n",
            "    6        -12.08297\n",
            "    7        -12.02586\n",
            "    8        -11.96490\n",
            "    9        -11.90346\n",
            "   10        -11.84533\n",
            "   11        -11.79443\n",
            "   12        -11.75444\n",
            "   13        -11.72836\n",
            "   14        -11.71822\n",
            "   15        -11.72474\n",
            "   16        -11.74731\n",
            "   17        -11.78402\n",
            "   18        -11.83197\n",
            "   19        -11.88758\n",
            "   20        -11.94703\n",
            "   21        -12.00666\n",
            "   22        -12.06329\n",
            "   23        -12.11440\n",
            "   24        -12.15830\n",
            "   25        -12.19407\n",
            "   26        -12.22142\n",
            "   27        -12.24053\n",
            "   28        -12.25192\n",
            "   29        -12.25628\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "<Figure size 432x288 with 1 Axes>"
            ],
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYQAAAD5CAYAAAAndkJ4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3deXhU5f3+8fcn+0IghISwBUIS9h0CisoqWFwRKO7WpRZEcWv7tbZurbWt+08rCqJWrVJEVNRKFQGRRUUIe8KWBAgJAZIACYQQsszz+yOjpZoFmEmeWT6v68qVyZk5mftcA9w85zznHDHGoJRSSgXYDqCUUsozaCEopZQCtBCUUko5aSEopZQCtBCUUko5aSEopZQCIMiVlUVkMvBHoAcwxBiT5lzeCngfGAy8aYyZXsf684Buzh+jgWJjTP+G3jc2NtYkJia6El0ppfzOunXriowxcXU971IhAOnAROCVHy0vBx4Geju/amWMufr7xyLyLFByOm+amJhIWlraGYdVSil/JiI59T3vUiEYY7Y53+THy48Dq0Qk5XR+j9T8gquA0a7kUUopdfY85RjCMOCgMSbTdhCllPJXDY4QRGQJ0KaWpx40xnzsphzXAnMbyDEFmALQsWNHN72tUkqp7zVYCMaYMY0ZQESCqDkOMaiBHLOB2QCpqal6ASallHIzT9hlNAbYbozJsx1EKaX8mUuFICITRCQPGAosFJFFpzy3B3gOuFlE8kSkp3P5ayKSesqvuYYGdhcppZRqfK7OMloALKjjucQ6lt/2o59vdiWDUkop93D1PASl/M6JimqWbDtIUelJBnRsSa92zQkO9IS9r0q5RgtBqdNQVe1gVVYRH2/MZ1HGAcoqqn94Liw4gL4doknt1JJBnVoysGNLWkaGWEyr1NnRQlCqDsYYNuYW8/HGfD7dnE9RaQVRYUFc3rcd4we0o3NsJOtzilmXc4R1OYeZvWIXVY6aCXDJcZEMchbE4MQYkuKaWd4apRom3ngLzdTUVKOXrlCNZVdhKR9tzOeTjfvYc6iMkKAALuzemvH92zOqexyhQYG1rneioprNecWk5Rxhfc4R1u09QnFZJQDXDE7gj1f0Iiy49nWVagoiss4Yk1rX8zpCUMop59Bx7pu3kfV7ixGBoUmtuGNkCj/r3YYW4cENrh8eEsg5Sa04J6kVAA6HYVfRcean5fLKil1sO3CMmdcPpF10eGNvilJnRUcISgEZ+SXc9I+1VDkcTB+VwuX92hHfPMxtv//z9AP8dv4mQoMCePG6AZyXHOu2363U6WpohKBTI5Tf+zb7ENe8spqQQOH924dy27Akt5YBwLjebfjozvOJjgjmxtfX8NrKXXjjf8aUb9NCUH7t8/QD3PTGGuJbhPH+tPNIaR3VaO+V0roZH0+/gLE94nl84TbumruBsoqqRns/pc6UFoLyW++u2csdc9bRq11z5k8d2iT79puFBjHzhoH8blx3/rNlPxNe+oY9Rccb/X2VOh1aCMrvGGN4aVkWD3y4hWFd4phz2zlNet6AiDBtZDJv3TqEg8fKuXzGKr7cfrDJ3l+pumghKL/icBge+3QrTy/awZX92/HaTalEhNiZbDesSxz/nn4BHWMiuPXNNJ5fshOHQ48rKHu0EJTfqKhycN97G3nj6z3cen5nnruqv/VLTiTERPDBtPOYOLA9zy/J5K65G7QUlDV6HoLyC2UVVdz+znpW7Czk/nHdmDYi+Se3frUlLDiQZyf3o0vrKJ78fDvd20Rx14VdbMdSfkgLQfm8I8cruOXNtWzOK+aJiX24Zojn3XFPRLh9RBI7Dx7juSU76ZsQzYiucbZjKT+ju4yUT6uqdvCrf6axdf9RZt4wyCPL4Hsiwl8n9KFbfBT3vLuBvCNltiMpP6OFoHzaS8uyScs5wlOT+vKzXrXdGtyzhIcEMvOGQVRXG+6Ys57yyuqGV1LKTbQQlM9al3OEv3+ZyZX923HlgPa245y2zrGRPHtVPzbnlfCnf2+1HUf5ES0E5ZOOlVdy77wNtG0RxmNX9rYd54xd1KsNd4xMZu6avbyXlms7jvITWgjKJz3ycQb5xeW8cE1/moc1fKVST/Sbi7pxfkorHv4onfR9JbbjKD+ghaB8zscb97Fgwz7uGp3CoE4xtuOctcAA4e/XDCAmMoRpc9ZRXFZhO5LycVoIyqfkHi7joQXppHZqyfRRKbbjuKxVs1Bevn4gB0rKuW/eRj1pTTUqLQTlM6qqHdw3byMA/+/q/gT5yI3vB3RsySOX92LZjkJe/DLLdhzlw3zjb4xS/HeK6Z+v7E1CTITtOG51wzkdmTigPc8v3clXOwpsx1E+SgtB+QRvnWJ6ukSEvzhPWrt33kZyD+tJa8r9tBCU1/P2KaanKzwkkFk3DKLaYZg2Z52etKbcTgtBeT1fmGJ6uhJjI3nuqv6k7zvKDD2eoNxMC0F5NV+ZYnomxvaM58r+7Zi9YpfebU25lRaC8lq+NsX0TPz+kh4EBwqPfaqXtlDu41IhiMhkEckQEYeIpJ6yvJWILBORUhGZUc/6/UVktYhsFJE0ERniSh7lP6odxienmJ6u+OZh3DumK19uL2DJVr39pnIPV/8WpQMTgRU/Wl4OPAz8toH1nwL+ZIzpDzzi/FmpBs1PyyUt5wh/Gt/L56aYnq6bz08kpXUz/vRphh5gVm7hUiEYY7YZY3bUsvy4MWYVNcVQ768AmjsftwDyXcmj/EPpySqe+WInqZ1aMsEHp5ieruDAAP50RS9yD5/gleW7bMdRPsD2OPte4GkRyQWeAX5f1wtFZIpzt1JaYWFhkwVUnmfWV9kUlZ7kwUt7eMxtMG05PyWWS/u05eWvsvTcBOWyBgtBRJaISHotX+Pd8P7TgPuMMQnAfcDrdb3QGDPbGJNqjEmNi9NbC/qr/OITvLpyF+P7t2NAx5a243iEBy/tQYAIjy/UA8zKNQ3eU9kYM6YR3/8m4B7n4/nAa434XsoHPL1oBwb4v591sx3FY7SLDmf66BSeXrSD5TsL9V7M6qzZ3mWUD4xwPh4NZFrMojzcptxiFmzYx20XdKZDS/88kFyX24Z1pnNsJH/8JIOTVXqAWZ0dV6edThCRPGAosFBEFp3y3B7gOeBmEckTkZ7O5a+dMkX1V8CzIrIJ+CswxZU8yncZY/jLwm3ENgth2shk23E8TmhQII9e3pPdRcd5fdVu23GUl2pwl1F9jDELgAV1PJdYx/LbTnm8ChjkSgblHxZlHGDNnsP8ZUJvonz88hRna2S31lzUM54Xl2YxYUB72rYItx1JeRnbu4yUalBFlYO/fbadrvHNuDo1wXYcj/bwZT1xGMPjC7fZjqK8kBaC8nj//HYPOYfKePDSnn53RvKZSoiJ4I6RKSzcvJ9vsopsx1FeRv92KY925HgFf1+ayfCucTp75jRNHZFEQkw4j36SQWW1w3Yc5UW0EJRHe2FpJqUnq3jwkh62o3iNsOBAHr2sF5kFpbz1zR7bcZQX0UJQHmtXYSnvrM7hmiEd6dYmynYcr3Jhj9aM6hbH80syKTja0BVklKqhhaA81t8+205YcCD3jelqO4rXEREevbzXDwfklTodWgjKI32bfYjFWw8ybWQycVGhtuN4pcTYSG69oDMfbdzH9gNHbcdRXkALQXkch8Pw+MKttI8O55cXdLYdx6vdPiKJZiFB/L/FO21HUV5AC0F5nA837CMj/yj3j+tGWHCg7TheLToihF8O68yijINsySuxHUd5OC0E5VHKKqp4etF2+iVEc0W/drbj+IRbL+hMdEQwzy3+ya1LlPofWgjKo7y2cjcHj57kYb3Xgds0DwtmyvAklu0oZF3OEdtxlAfTQlAe42h5Ja+u3MVFPeNJTYyxHcen3HxeIrHNQnSUoOqlhaA8xj+/2cOx8iruvrCL7Sg+JyIkiNtHJPN11iG+zT5kO47yUFoIyiOUnqzitVW7GdOjNb3bt7AdxyfdcG4n4puH8tziHRhjbMdRHkgLQXmEt7/NobiskrtG6+igsYQFBzJ9VApr9xxhZaZe+E79lBaCsq6soopXV+5iRNc4+iVE247j064anED76HCe/UJHCeqntBCUdXNW7+Xw8Qo9dtAEQoMCufvCFDbllbB0W4HtOMrDaCEoq8orq3llxS4uSIllUKeWtuP4hYkDO5DYKoJnF+/E4dBRgvovLQRl1dw1eykqPamjgyYUHBjAPWO6sG3/UT7POGA7jvIgWgjKmvLKamYtz+bcpBiGdNbzDprSFf3ak9K6Gc8t3km1jhKUkxaCsmZ+Wi4Hj57kbp1Z1OQCA4T7xnQlq6CUf2/Ktx1HeQgtBGVFRZWDmV9lk9qpJUOTW9mO45cu7t2GHm2b8/ySnVTprTYVWgjKkg/W55FfUs7dF3bRaxZZEhAg/HpsV/YcKuPD9ftsx1EeQAtBNbnKagcvLcuif0I0w7rE2o7j18b0aE2/Di14YWkmJ6uqbcdRlmkhqCa3YMM+8o6c4B4dHVgnIvz6om7sKz7Be2tzbcdRlmkhqCZV5Rwd9GnfgpHd4mzHUcDwLrEMTmzJjGVZlFfqKMGfaSGoJvXJpnxyDpVx1+gUHR14CBHh12O7cfDoSeau2Ws7jrJIC0E1mWqHYcayLHq0bc7YnvG246hTDE1uxZDEGGav2EVFlc448lcuFYKITBaRDBFxiEjqKctbicgyESkVkRn1rN9PRL4VkS0i8m8Rae5KHuXZFm7Zz67C49ytowOPdOfoFPaXlPPh+jzbUZQlro4Q0oGJwIofLS8HHgZ+28D6rwEPGGP6AAuA/3Mxj/JQDofhxaWZdI1vxs96tbEdR9VieJdY+rRvwczl2Xpegp9yqRCMMduMMT+5J58x5rgxZhU1xVCfrvy3TBYDk1zJozzX5xkHyCwo5a7RXQgI0NGBJxIR7hyVQs6hMhZu2W87jrLA9jGEDGC88/FkIKGuF4rIFBFJE5G0wsLCJgmn3MPhMPx9aSZJcZFc0qet7TiqHhf1jKdrfDNeWpalV0L1Qw0WgogsEZH0Wr7GN7TuabgVuENE1gFRQEVdLzTGzDbGpBpjUuPidLqiN1m2o4DtB44xfVQKgTo68GgBAcIdI1PYebCUxdsO2o6jmlhQQy8wxoxprDc3xmwHLgIQka7ApY31XsqeWcuzaR8dzhX92tmOok7DZX3b8tzinby0LIuLesbrBAA/YnWXkYi0dn4PAB4CZtnMo9xvXc5h1u45wq+GdSYo0PYeSnU6ggIDmDYymc15JXrvZT/j6rTTCSKSBwwFForIolOe2wM8B9wsInki0tO5/LVTpqheKyI7ge1APvCGK3mU55n51S5aRgRz1eA6Dw8pDzRxYHvaNA9jxrIs21FUE2pwl1F9jDELqJkuWttziXUsv+2Uxy8AL7iSQXmunQePsWTbQe4d04WIEJf+qKkmFhoUyJThSTz26VbW7jnM4ES9gZE/0DG8ajSvLN9FeHAgNw1NtB1FnYVrh3SkVWQIM77UUYK/0EJQjSK/+AQfb9zH1YMTaBkZYjuOOgvhIYHcekFnlu8sZEteie04qgloIahG8fqq3RjgtmGdbUdRLrhxaCeiwoJ4SY8l+AUtBOV2xWUVzF2zlyv6taNDywjbcZQLmocFc/N5iTVnmh88ZjuOamRaCMrt3v42h7KKaqaOSLIdRbnBLed3Jjw4kJe/yrYdRTUyLQTlVuWV1bz5zR5GdYujexu9eK0viIkM4fpzOvLJpnz2HiqzHUc1Ii0E5Vbz03I5dLyC20ck246i3OhXw5MIFGHmch0l+DItBOU2VdUOZq/cxYCO0QzprPPWfUl88zAmp3bgg3V5HChp6CLGyltpISi3+U/6AXIPn+D2Ecl6/RsfdPuIZKqNYfaKXbajqEaihaDcwhjDrK+ySY6LZGwPvT2mL0qIiWB8/3b8a00Oh0pP2o6jGoEWgnKLlZlFbN1/lKnDk/UGOD7sjpEpnKxy8I+vd9uOohqBFoJyi1nLs4lvHsr4AXqJa1+W0roZF/duwz+/yaHkRKXtOMrNtBCUyzbnFfNN9iF+eUFnQoMCbcdRjeyOkSkcO1nFO6tzbEdRbqaFoFw2a3k2UWFBXDuko+0oqgn0bt+CUd3ieH3VbsoqqmzHUW6khaBcsrvoOJ+lH+DGczsRFRZsO45qIneOSuHw8Qrmrsm1HUW5kRaCcsnsFbsIDgzglvP1Inb+JDUxhnM6x/Dqil2crKq2HUe5iRaCOmsFx8r5YH0ePx/UgbioUNtxVBO7c1QKB46W8+H6fbajKDfRQlBn7Y2v91BV7WDKML2InT8a1iWWvh1aMGt5NlXVDttxlBtoIaizcrS8kne+zeHi3m1JjI20HUdZICLcOSqFnENlLNyy33Yc5QZaCOqszFm9l2Mnq5g2Ui9i58/G9oina3wzXlqWhcNhbMdRLtJCUGesvLKa11ftZliXWHq3b2E7jrIoIEC4Y2QKOw+WsmTbQdtxlIu0ENQZ+2B9HkWlJ3V0oAC4rG9bOsZE8NKyLIzRUYI300JQZ6Sq2sEry3fRLyGaoUmtbMdRHiAoMIDbRySzKa+EVVlFtuMoF2ghqDPyWfoB9h4uY5pe4lqdYtKg9sQ3D+WlZVm2oygXaCGo02aMYeZX2STFRXJRT73Etfqv0KBAfjUsidW7DrMu57DtOOosaSGo07bCeYnr20foJa7VT113TkdiIkOY8aWOEryVFoI6bTO/yqJN8zCu7N/edhTlgSJCgrj1/ESW7SgkI7/Edhx1FlwqBBGZLCIZIuIQkdRTlo8VkXUissX5fXQd68eIyGIRyXR+b+lKHtV4Nuw9wupdh7ltWGdCgvT/Eap2Nw5NJCo0iJeXZduOos6Cq3+z04GJwIofLS8CLjfG9AFuAt6uY/0HgKXGmC7AUufPygPNWp5Ni/BgvcS1qleL8GBuHNqJ/6TvJ6ug1HYcdYZcKgRjzDZjzI5alm8wxuQ7f8wAwkWktqufjQfecj5+C7jSlTyqcWQVHGNRxkFuOi+RyNAg23GUh7v1gs6EBgUwa7mOErxNU4z9JwHrjTG13ZU73hjz/UVQDgA6dcUDvbJ8F2HBAdx8XqLtKMoLxDYL5ZrBHflowz7yjpTZjqPOQIOFICJLRCS9lq/xp7FuL+BJYGpDrzU1pzjWeZqjiEwRkTQRSSssLGzo1yk3yS8+wUcb93HN4JoZJEqdjinDkxCp+c+E8h4NFoIxZowxpnctXx/Xt56IdAAWAL8wxtQ1djwoIm2dr28LFNSTY7YxJtUYkxoXF9dQbOUmr6/ajTFw2zC9AY46fe2iw5k4oAPz0nIpOFpuO446TY2yy0hEooGFwAPGmK/reekn1Bx0xvm93pJRTevI8QrmrtnLFf3b0aFlhO04ystMG5lMtcPwygodJXgLV6edThCRPGAosFBEFjmfmg6kAI+IyEbnV2vnOq+dMkX1CWCsiGQCY5w/Kw/xz29zKKuo5vYRehE7deYSYyO5sn973lmdo6MEL+HqLKMFxpgOxphQY0y8MeZnzuWPG2MijTH9T/kqcD53mzEmzfn4kDHmQmNMF+euKT3n3UOUVVTx5je7GdMjnq7xUbbjKC9194UpVDkMM3XGkVfQM4xUreatzeVIWaVe4lq5pFOrSCYOaM+c7/ZyoERHCZ5OC0H9RGW1g1dX7GJI5xgGddKTx5Vr7hrdBYfDMPMrvcaRp9NCUD/xycZ88kvKdXSg3KJjqwgmDezA3DW57C85YTuOqocWgvofDodh1vJsureJYmRXnd6r3GP66BQcxug1jjycFoL6H4syDpBZUMq0kXoDHOU+CTERTE5NYN7aXPKLdZTgqbQQ1A8cDsMLSzNJiovksr7tbMdRPmb66BQMRu+q5sG0ENQPFmUcYPuBY9xzYRcC9QY4ys3aR4dzVWoC76Xl6jWOPJQWggJ0dKCaxp2jUhCEl/RYgkfSQlCAjg5U02gXHc7VgxOYn5ZL7mEdJXgaLQSlowPVpO4YVXNPbr33sufRQlA6OlBNqm2LcK4b0pH31+ex95COEjyJFoKf09GBsmHayGSCAoQXv8y0HUWdQgvBz+noQNkQ3zyM687pyIcb9rGn6LjtOMpJC8GP6ehA2TRtxPejBD2W4Cm0EPzY5zo6UBa1bh7Gjed2YsGGPHbrKMEjaCH4KYfD8MKSTJJ1dKAsmjoimZCgAF5cqscSPIEWgp/6POMAOw4e424dHSiL4qJC+cXQRD7auI/swlLbcfyeFoIf0tGB8iRThicRGhTIC0t0lGCbFoIf0tGB8iSxzUK55fxEPtmUz+a8Yttx/JoWgp/R0YHyRNNGJtMqMoTHF27DGGM7jt/SQvAzOjpQnigqLJh7x3Zlze7DfLH1oO04fksLwY/o6EB5smsHJ5DSuhlPfLadiiqH7Th+SQvBj+joQHmyoMAA/nBJd3YXHWfOdzm24/glLQQ/oaMD5Q1GdWvN+SmteGFpJiVllbbj+B0tBD/xWbqODpTnExEevKQnJScqmbFMp6E2NS0EP1BR5eCZL3aQ0rqZjg6Ux+vZrjmTB3XgrW9y9PLYTUwLwQ+8vTqH3UXHefCSHjo6UF7hNxd1IzBAePLz7baj+BUtBB9XXFbB35dmMqxLLCO7xdmOo9RpiW8extQRSSzcsp91OYdtx/EbLhWCiEwWkQwRcYhI6inLx4rIOhHZ4vw++kzWV+7z/JJMjpVX8tClPRHR0YHyHlOGJxHfPJQ/f6onqzUVV0cI6cBEYMWPlhcBlxtj+gA3AW+f4frKDbILS3lndQ7XDOlItzZRtuModUYiQoL4zUXd2JhbzL8377cdxy+4VAjGmG3GmB21LN9gjMl3/pgBhItI6Omur9zjb//ZRlhwIPeN6Wo7ilJnZdLADvRo25wnP9tOeWW17Tg+rymOIUwC1htjTjbBeymnr7OKWLKtgDtHpRAX9ZMuVsorBAYID13ag33FJ3jzmz224/i8BgtBRJaISHotX+NPY91ewJPAVFeDisgUEUkTkbTCwkJXf51Pq3YY/vzpVjq0DOeW8xNtx1HKJeenxDK6e2te+jKLQ6X6/8rG1GAhGGPGGGN61/L1cX3riUgHYAHwC2NMtqtBjTGzjTGpxpjUuDidLVOf+Wm5bD9wjAcu7k5YcKDtOEq57A+XdKesspoX9M5qjapRdhmJSDSwEHjAGPN1Y7yHql3pySqe+WInqZ1acmmftrbjKOUWKa2juHZIAnO+20tWgd5ZrbG4Ou10gojkAUOBhSKyyPnUdCAFeERENjq/WjvXee37Kab1rK/O0syvsigqPclDl+k0U+Vb7h3TlfDgQJ74bJvtKD4ryJWVjTELqNkt9OPljwOP17HObQ2tr85O3pEyXl25myv7t6N/QrTtOEq5VWyzUO4YlcxTn+9gxc5ChnfVXcfupmcq+5CnPt9BgMD947rbjqJUo7j1/M4kxUbyhwVbOH6yynYcn6OF4CPW7z3CJ5vymTIsiXbR4bbjKNUowoIDeWJSX/KOnODpRXoKk7tpIfgAY2qmmbaOCmXqiGTbcZRqVEM6x/CLoZ1469s9ep0jN9NC8AH/3ryfDXuL+e3PuhEZ6tJhIaW8wv3jutOuRTj3v79Zz2B2Iy0EL1deWc2Tn22nZ9vmTBrYwXYcpZpEs9Ag/jqxD9mFx5nxZZbtOD5DC8HLvb5qN/uKT/DwZT31XgfKr4zoGsekgR2YuTybjPwS23F8ghaCF8svPsHLy7K4qGc8Q5Nb2Y6jVJN7+LIetIwI4f73N1NV7bAdx+tpIXgpYwy/+2AzBnjo0p624yhlRXRECH8e34uM/KPMXrnLdhyvp4XgpeZ8t5eVmUX84ZIedGwVYTuOUtZc3KctF/duw/NLMsku1MtauEILwQvlHDrOX/+zjWFdYrn+nI624yhl3Z/G9yI8OJAHPtiMw6F3VztbWgheptph+L/5m2tuQD6pr16vSCmgdVQYD1/Wk7V7jvD26hzbcbyWFoKXeePr3azZc5g/Xt5Lz0hW6hSTBrZneNc4nvx8O3lHymzH8UpaCF4kq+AYTy3awdie8Uwc2N52HKU8iojw1wm9Afj9h1swRncdnSktBC9RVe3g1+9tIjIkkL9O6KO7ipSqRYeWEfxuXHdWZhbxwfp9tuN4HS0ELzHzq2w255Xwlwl99B7JStXjxnM7MTixJX/+dCsFx8ptx/EqWgheICO/hBeWZnJFv3ZcondBU6peAQHCE5P6cqKymj98mK67js6AFoKHO1lVzW/e20TLyBAeG9/LdhylvEJyXDN+N647S7Yd5OWvXL6lu9/QQvBwLyzJZPuBYzw5qQ/RESG24yjlNW49P5Hx/dvxzBc7WLajwHYcr6CF4MHW7z3CrOXZXJ2awOju8bbjKOVVRIQnJvale5vm3DN3A3uKjtuO5PG0EDzUiYpqfvveJtq2COehy3rYjqOUVwoPCWT2jYMICBCmvJ2mt91sgBaCh3pq0XZ2FR3n6Z/3JSos2HYcpbxWQkwEM64dSFZBKf/3/iY9yFwPLQQP9E12EW98vYebz0vkvJRY23GU8noXdInlgYu7858tB5i5XA8y10ULwcPsKizlzjnrSYqN5P5x3WzHUcpn/GpYEpf3a8fTi3bwlR5krpUWggcpPHaSm95YQ4AI/7h5MBEhen9kpdxFRHhyUh+6xUdx99wN5BzSg8w/poXgIY6frOKXb62l8NhJXr95MImxkbYjKeVzIkKCmH1jKiLClH+u04PMP6KF4AGqqh1M/9d60veV8NJ1A+mfEG07klI+q2OrCGZcN4DMgmPc//5mPch8Ci0Ey4wxPPxxOst2FPLnK3tzYQ8930CpxjasSxz3j+vOwi37mbVcb735PS0Ey178Mou5a3KZPiqF68/pZDuOUn5j6vAkLu3blqcWbWf5zkLbcTyCS4UgIpNFJENEHCKSesrysSKyTkS2OL+PrmP9p0Vku4hsFpEFIuJX+0rmp+Xy3OKdTBzYnt9c1NV2HKX8iojw9M/7/nCQOavgmO1I1rk6QkgHJgIrfrS8CLjcGNMHuAl4u471FwO9jTF9gZ3A713M4zWW7yzk9x9uYViXWJ6YqLfCVMqG7w8yBwcGcPUrq9maf9R2JKtcKgRjzHIvmaMAAAn1SURBVDZjzI5alm8wxuQ7f8wAwkXkJxfxN8Z8YYz5/jD/aqCDK3m8Rfq+Eu54Zx1d46N4+fqBhATpnjulbOnYKoJ5U88lODCAa19dzcbcYtuRrGmKf4kmAeuNMScbeN2twGd1PSkiU0QkTUTSCgu9d39f7uEybnlzLdERIbxxy2C9LIVSHiA5rhnzbx9K8/AgbnjtO9bsPmw7khUNFoKILBGR9Fq+xp/Gur2AJ4GpDbzuQaAKmFPXa4wxs40xqcaY1Li4uIbe2iMVl1Vw8xtrOFlZzZu3DCa+eZjtSEopp4SYCOZPPY/WzUP5xT++Y2Wm9/7H82w1WAjGmDHGmN61fH1c33oi0gFYAPzCGFPnxUNE5GbgMuB648MTgkvKKrntrTRyD5/gtZsG0yU+ynYkpdSPtGkRxrwpQ0lsFckv30xjydaDtiM1qUbZZeScLbQQeMAY83U9rxsH3A9cYYwpa4wsniB9XwmXzVjJprxinr+mP0M6x9iOpJSqQ1xUKO9OOZcebaO4/Z11fLo5v+GVfISr004niEgeMBRYKCKLnE9NB1KAR0Rko/OrtXOd106ZojoDiAIWO18zy5U8nmje2r1MnPkNVdWGeVOH6j2RlfIC0REhvHPbOQzoGM3dczfw/ro825GahHjjXprU1FSTlpZmO0a9yiurefijdOavy+OClFheuKY/rZr9ZKKVUsqDlVVUMfXtdazMLOLPV/bmxnO9++RREVlnjEmt63m9nGYj2FN0nGlz1rNt/1HuHp3CPWO6Ehig5xko5W0iQoJ49RepTP/Xeh7+KJ3yimp+NTzJdqxGo4XgZl9kHOA38zcRIMIbNw9mVPfWtiMppVwQFhzIzBsGce+8jfzlP9s4XFbBr8d2JTjQ984f0kJwk6pqB898sZNZy7Pp074FL18/kISYCNuxlFJuEBwYwN+vGUDzsCBmfpXNysxCnpncj+5tmtuO5la+V3EWFB47yQ2vf8es5dlcd05H5t8+VMtAKR8TGCD8bWJfZt0wiAMl5Vz+4ipmfJlJVbXDdjS30RGCC6qqHSzeepBHP8ngaHklz07ux6RBfnH1DaX81rjebRjSOYZHP8ngmS92sijjIM9M7ke3Nt5/bpHOMjoLBUfLeXdtLv/6bi8HjpaTFBfJS9cNpEdb3xo+KqXq99mW/Tz0UTrHyqu4Z0wXpg5PIsiDjy3oLCM3Mcbw3e7DvL06h0XpB6hyGIZ3jeOx8b0Y3b21R/8hUEo1jov7tGVI5xge+SSDpxft4IuMAzwzuZ/XXolARwgNOFpeyYL1+3hndQ6ZBaW0CA/mqtQOXHdOJzrrfY+VUk4LN+/n4Y/TKS2v4t6xXZgyzPNGCzpCOEPlldWUnKhkX/EJ3l+Xx0cb9lFWUU2/Di14+ud9ubxfO8KCA23HVEp5mEv7tuWcpBge+Tidpz7fwaL0A9wzpgsjurb2mvOQ/GqEsGb3YTbnFXOkrILiskqKyyo5UlbBkbJKSpzfT1RW//D60KAArujXjhvO7UQ/vfG9Uuo0fbo5n8f+vZWCYydp1yKMqwYncFVqAu2iw63mamiE4FeF8OjH6bz1bQ6BAUJ0eDDREcG0jAghOiKY6IgQosODaRkZ8sPy85JbER0R0ghboJTydZXVDpZuO8ic7/ayMrOIAIFR3Vpz7ZCOjOwWZ2V3khbCKUpOVCICUaFBestKpVSTyT1cxrtr9/JeWh6Fx07SpnnNqOHqwQm0b8JRgxaCUkp5iJpRQwFz1+xlRWYhAozs1ppxvduQHNeM5LjIRt0roQeVlVLKQwQHBjCudxvG9W5D7uEy3kvLZd7aXL7cXvDDa2IiQ0iKjSQpLpKkuGbOx83o1Cqi0a+fpCMEpZSyqNphyDl0nN1Fx9lVeJxdRaVkF9Y8Lir9763oAwOEjjER/HVCH4Ymtzqr99IRglJKebDAAKkZCcQ148Ie//tcyYlKZ1GU/lAWrZo13i4lLQSllPJQLcKD6Z8QTf8mmvbuWafRKaWUskYLQSmlFKCFoJRSykkLQSmlFKCFoJRSykkLQSmlFKCFoJRSykkLQSmlFOCll64QkUIg5yxXjwWK3BjHE/jaNvna9oDvbZOvbQ/43jbVtj2djDFxda3glYXgChFJq+9aHt7I17bJ17YHfG+bfG17wPe26Wy2R3cZKaWUArQQlFJKOfljIcy2HaAR+No2+dr2gO9tk69tD/jeNp3x9vjdMQSllFK188cRglJKqVr4VSGIyDgR2SEiWSLygO08rhKRPSKyRUQ2iohX3kJORP4hIgUikn7KshgRWSwimc7vLW1mPBN1bM8fRWSf83PaKCKX2Mx4pkQkQUSWichWEckQkXucy73yc6pne7z2cxKRMBFZIyKbnNv0J+fyziLynfPfvHkiUu/ddfxml5GIBAI7gbFAHrAWuNYYs9VqMBeIyB4g1RjjtXOnRWQ4UAr80xjT27nsKeCwMeYJZ3G3NMb8zmbO01XH9vwRKDXGPGMz29kSkbZAW2PMehGJAtYBVwI344WfUz3bcxVe+jmJiACRxphSEQkGVgH3AL8GPjTGvCsis4BNxpiZdf0efxohDAGyjDG7jDEVwLvAeMuZ/J4xZgVw+EeLxwNvOR+/Rc1fVq9Qx/Z4NWPMfmPMeufjY8A2oD1e+jnVsz1ey9Qodf4Y7PwywGjgfefyBj8jfyqE9kDuKT/n4eV/CKj5wL8QkXUiMsV2GDeKN8bsdz4+AMTbDOMm00Vks3OXklfsWqmNiCQCA4Dv8IHP6UfbA178OYlIoIhsBAqAxUA2UGyMqXK+pMF/8/ypEHzRBcaYgcDFwJ3O3RU+xdTs0/T2/ZozgWSgP7AfeNZunLMjIs2AD4B7jTFHT33OGz+nWrbHqz8nY0y1MaY/0IGaPSLdz/R3+FMh7AMSTvm5g3OZ1zLG7HN+LwAWUPOHwBccdO7n/X5/b4HlPC4xxhx0/mV1AK/ihZ+Tc7/0B8AcY8yHzsVe+znVtj2+8DkBGGOKgWXAUCBaRIKcTzX4b54/FcJaoIvzqHsIcA3wieVMZ01EIp0HxBCRSOAiIL3+tbzGJ8BNzsc3AR9bzOKy7//RdJqAl31OzgOWrwPbjDHPnfKUV35OdW2PN39OIhInItHOx+HUTJ7ZRk0x/Nz5sgY/I7+ZZQTgnEb2PBAI/MMY8xfLkc6aiCRRMyoACAL+5Y3bIyJzgZHUXJnxIPAo8BHwHtCRmqvaXmWM8YoDtXVsz0hqdkMYYA8w9ZR97x5PRC4AVgJbAIdz8R+o2e/udZ9TPdtzLV76OYlIX2oOGgdS8x/994wxjzn/nXgXiAE2ADcYY07W+Xv8qRCUUkrVzZ92GSmllKqHFoJSSilAC0EppZSTFoJSSilAC0EppZSTFoJSSilAC0EppZSTFoJSSikA/j+XEejGlRgeXgAAAABJRU5ErkJggg==\n"
          },
          "metadata": {
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Como agora estamos a considerar menos átomos, os cálculos são muito mais rápidos... também é possível que precisemos de parametros menos agressivos no cálculo `--path`."
      ],
      "metadata": {
        "id": "4Ix1d0lUiMmi"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# calculo path\n",
        "\n",
        "p1_fn = '/content/work_dir/reac01-lmm-teste-lmm000-guess.xyz'\n",
        "p2_fn = '/content/work_dir/reac01-lmm-teste-lmm025-guess.xyz'\n",
        "\n",
        "!rm reac01-path*\n",
        "\n",
        "# ver https://github.com/grimme-lab/xtb/blob/main/man/xcontrol.7.adoc\n",
        "\n",
        "with open('xtb-path.conf','w') as f:\n",
        "  f.write(\"\"\"\n",
        "\n",
        "$path\n",
        " # number of runs for pathfinder\n",
        " nrun = 5\n",
        " # number of points on the path to optimize\n",
        " #nopt = 20\n",
        " npoint = 20\n",
        " # number of steps to optimize the points on the path\n",
        " anopt = 1\n",
        " # factor for RMSD criterium pushing away from the reactant structure\n",
        " kpull = -0.1\n",
        " # factor for RMSD criterium pulling towards the product structure\n",
        " kpush = 0.01\n",
        " # width of the RMSD criterium\n",
        " alp = 0.001\n",
        "$end\n",
        "  \"\"\")\n",
        "\n",
        "run_xtb(f\"{p1_fn} --path {p2_fn} --input xtb-path.conf --namespace reac01-path > reac01-path.out\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KIz8eGcitIEZ",
        "outputId": "ed679f04-5288-46bf-902d-59e3631ab08b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "b''"
            ]
          },
          "metadata": {},
          "execution_count": 112
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "! rm reac01-ts-hess*\n",
        "run_xtb(\"/content/work_dir/reac01-path.xtbpath_ts.xyz --hess --acc 0.001 --namespace reac01-ts-hess > reac01-ts-hess.out\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "5cXMyAIbtzM5",
        "outputId": "4dbff4bf-69d8-451b-ec47-b089da7f45f7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "b''"
            ]
          },
          "metadata": {},
          "execution_count": 113
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "with open('xtb-opt.conf','w') as f:\n",
        "  f.write(\"\"\"\n",
        "$opt\n",
        " engine=inertial\n",
        " optlevel=-1\n",
        " rcut=1.2\n",
        " maxdispl=0.00001\n",
        " microcycle=1\n",
        "$end\n",
        "  \"\"\")\n",
        "\n",
        "run_xtb(f\"{p1_fn} --ohess --acc 0.00001 --namespace reac01-reactant-hess  > reac01-reactant-hess.out\")\n",
        "run_xtb(f\"{p2_fn} --ohess --acc 0.00001 --namespace reac01-product-hess  > reac01-product-hess.out\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "IDxFBzq03EY2",
        "outputId": "00ad6f1f-cd0b-4d41-e2e3-32ee04d3ba20"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "b''"
            ]
          },
          "metadata": {},
          "execution_count": 170
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "compile_reaction_data('reac01-reactant-hess.out','reac01-ts-hess.out','reac01-product-hess.out')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "nPgme35k-Muy",
        "outputId": "12511247-5742-487b-9aea-6a2eba9968c9"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581897372909281\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581897372909281\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581897372909281 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581897372909281\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581897372909281 = $3Dmol.createViewer($(\"#3dmolviewer_16581897372909281\"),{backgroundColor:\"white\"});\nviewer_16581897372909281.zoomTo();\n\tviewer_16581897372909281.addModel(\"4\\n Created by Molecule Class\\nC          -0.101887        -0.568042         0.343563\\nO           0.201680        -1.697471         0.470350\\nC           0.102777         0.568574        -0.341872\\nO          -0.202570         1.696939        -0.472041\\n\");\n\tviewer_16581897372909281.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581897372909281.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581897372909281.zoomTo();\nviewer_16581897372909281.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581897372909281\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581897372909281\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581897372909281 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581897372909281\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581897372909281 = $3Dmol.createViewer($(\"#3dmolviewer_16581897372909281\"),{backgroundColor:\"white\"});\n",
              "viewer_16581897372909281.zoomTo();\n",
              "\tviewer_16581897372909281.addModel(\"4\\n Created by Molecule Class\\nC          -0.101887        -0.568042         0.343563\\nO           0.201680        -1.697471         0.470350\\nC           0.102777         0.568574        -0.341872\\nO          -0.202570         1.696939        -0.472041\\n\");\n",
              "\tviewer_16581897372909281.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581897372909281.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581897372909281.zoomTo();\n",
              "viewer_16581897372909281.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reactant final geometry\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581897372948122\"  style=\"position: relative; width: 400px; height: 400px\">\n        <p id=\"3dmolwarning_16581897372948122\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581897372948122 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581897372948122\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581897372948122 = $3Dmol.createViewer($(\"#3dmolviewer_16581897372948122\"),{backgroundColor:\"white\"});\nviewer_16581897372948122.zoomTo();\n\tviewer_16581897372948122.addModel(\"4\\n* (null), Energy   -1000.0000000\\nC -0.068569 -1.268235 1.145713 -0.09 -0.71 0.49\\nO 0.034670 -1.858697 0.190507 0.03 0.25 -0.08\\nC 0.358137 1.185731 -0.431866 0.13 0.22 -0.26\\nO -0.351793 1.910778 -0.925401 -0.06 0.17 -0.12\\n\",\"xyz\",{\"vibrate\": {\"frames\": 10, \"amplitude\": 1}});\n\tviewer_16581897372948122.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581897372948122.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581897372948122.animate({\"loop\": \"backAndForth\"});\n\tviewer_16581897372948122.zoomTo();\nviewer_16581897372948122.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581897372948122\"  style=\"position: relative; width: 400px; height: 400px\">\n",
              "        <p id=\"3dmolwarning_16581897372948122\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581897372948122 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581897372948122\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581897372948122 = $3Dmol.createViewer($(\"#3dmolviewer_16581897372948122\"),{backgroundColor:\"white\"});\n",
              "viewer_16581897372948122.zoomTo();\n",
              "\tviewer_16581897372948122.addModel(\"4\\n* (null), Energy   -1000.0000000\\nC -0.068569 -1.268235 1.145713 -0.09 -0.71 0.49\\nO 0.034670 -1.858697 0.190507 0.03 0.25 -0.08\\nC 0.358137 1.185731 -0.431866 0.13 0.22 -0.26\\nO -0.351793 1.910778 -0.925401 -0.06 0.17 -0.12\\n\",\"xyz\",{\"vibrate\": {\"frames\": 10, \"amplitude\": 1}});\n",
              "\tviewer_16581897372948122.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581897372948122.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581897372948122.animate({\"loop\": \"backAndForth\"});\n",
              "\tviewer_16581897372948122.zoomTo();\n",
              "viewer_16581897372948122.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Animation of the Reaction Coord. at the TS\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581897372994432\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581897372994432\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581897372994432 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581897372994432\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581897372994432 = $3Dmol.createViewer($(\"#3dmolviewer_16581897372994432\"),{backgroundColor:\"white\"});\nviewer_16581897372994432.zoomTo();\n\tviewer_16581897372994432.addModel(\"4\\n Created by Molecule Class\\nC           0.053355        -0.519118         0.423102\\nO          -0.497209        -1.187230         1.219572\\nC          -0.050587         0.518990        -0.421455\\nO           0.494442         1.187359        -1.221219\\n\");\n\tviewer_16581897372994432.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581897372994432.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581897372994432.zoomTo();\nviewer_16581897372994432.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581897372994432\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581897372994432\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581897372994432 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581897372994432\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581897372994432 = $3Dmol.createViewer($(\"#3dmolviewer_16581897372994432\"),{backgroundColor:\"white\"});\n",
              "viewer_16581897372994432.zoomTo();\n",
              "\tviewer_16581897372994432.addModel(\"4\\n Created by Molecule Class\\nC           0.053355        -0.519118         0.423102\\nO          -0.497209        -1.187230         1.219572\\nC          -0.050587         0.518990        -0.421455\\nO           0.494442         1.187359        -1.221219\\n\");\n",
              "\tviewer_16581897372994432.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581897372994432.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581897372994432.zoomTo();\n",
              "viewer_16581897372994432.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reactant final geometry\n",
            "                Energy          Enthalpy       Gibbs Energy  \n",
            " Delta (D)             0.00             0.01             0.01\n",
            " Eact (D)            155.21           150.25           150.25\n",
            " Delta (I)            -0.00            -0.01            -0.01\n",
            " Eact (I)            155.20           150.24           150.24\n",
            "Value of the imaginary frequency at TS:              -100.33\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# No caso anterior, os reagentes \"saltam\" a barreira e optimizam para os produtos\n",
        "# neste caso, temos que ir buscar uma estrutura inicial dos reagentes a uma frame anterior.\n",
        "\n",
        "extract_geo('/content/work_dir/sys03-mtd03.xtb.trj',33,'reac1-reactant-preguess.xyz')\n",
        "isolate_molecules_with_atoms('reac1-reactant-preguess.xyz',[123-1,125-1],'reac1-reactant-guess.xyz')\n",
        "! rm reac01-reactant-hess*\n",
        "run_xtb(f\"reac1-reactant-guess.xyz --ohess --acc 0.00001 --namespace reac01-reactant-hess  > reac01-reactant-hess.out\")\n",
        "\n",
        "# e agora sim temos um relatório da reacção elementar\n",
        "compile_reaction_data('reac01-reactant-hess.out','reac01-ts-hess.out','reac01-product-hess.out')\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "nwxUhs_ut1nj",
        "outputId": "b25e0bb3-a485-46e7-f902-05d2bec84fe0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581897837026763\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581897837026763\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581897837026763 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581897837026763\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581897837026763 = $3Dmol.createViewer($(\"#3dmolviewer_16581897837026763\"),{backgroundColor:\"white\"});\nviewer_16581897837026763.zoomTo();\n\tviewer_16581897837026763.addModel(\"4\\n Created by Molecule Class\\nC          -0.259357        -1.995755         0.272956\\nO           0.426264        -1.167297        -0.066030\\nC           0.383766         1.427429         0.172359\\nO          -0.550673         1.735623        -0.379283\\n\");\n\tviewer_16581897837026763.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581897837026763.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581897837026763.zoomTo();\nviewer_16581897837026763.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581897837026763\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581897837026763\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581897837026763 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581897837026763\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581897837026763 = $3Dmol.createViewer($(\"#3dmolviewer_16581897837026763\"),{backgroundColor:\"white\"});\n",
              "viewer_16581897837026763.zoomTo();\n",
              "\tviewer_16581897837026763.addModel(\"4\\n Created by Molecule Class\\nC          -0.259357        -1.995755         0.272956\\nO           0.426264        -1.167297        -0.066030\\nC           0.383766         1.427429         0.172359\\nO          -0.550673         1.735623        -0.379283\\n\");\n",
              "\tviewer_16581897837026763.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581897837026763.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581897837026763.zoomTo();\n",
              "viewer_16581897837026763.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reactant final geometry\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581897837089605\"  style=\"position: relative; width: 400px; height: 400px\">\n        <p id=\"3dmolwarning_16581897837089605\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581897837089605 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581897837089605\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581897837089605 = $3Dmol.createViewer($(\"#3dmolviewer_16581897837089605\"),{backgroundColor:\"white\"});\nviewer_16581897837089605.zoomTo();\n\tviewer_16581897837089605.addModel(\"4\\n* (null), Energy   -1000.0000000\\nC -0.068569 -1.268235 1.145713 -0.09 -0.71 0.49\\nO 0.034670 -1.858697 0.190507 0.03 0.25 -0.08\\nC 0.358137 1.185731 -0.431866 0.13 0.22 -0.26\\nO -0.351793 1.910778 -0.925401 -0.06 0.17 -0.12\\n\",\"xyz\",{\"vibrate\": {\"frames\": 10, \"amplitude\": 1}});\n\tviewer_16581897837089605.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581897837089605.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581897837089605.animate({\"loop\": \"backAndForth\"});\n\tviewer_16581897837089605.zoomTo();\nviewer_16581897837089605.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581897837089605\"  style=\"position: relative; width: 400px; height: 400px\">\n",
              "        <p id=\"3dmolwarning_16581897837089605\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581897837089605 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581897837089605\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581897837089605 = $3Dmol.createViewer($(\"#3dmolviewer_16581897837089605\"),{backgroundColor:\"white\"});\n",
              "viewer_16581897837089605.zoomTo();\n",
              "\tviewer_16581897837089605.addModel(\"4\\n* (null), Energy   -1000.0000000\\nC -0.068569 -1.268235 1.145713 -0.09 -0.71 0.49\\nO 0.034670 -1.858697 0.190507 0.03 0.25 -0.08\\nC 0.358137 1.185731 -0.431866 0.13 0.22 -0.26\\nO -0.351793 1.910778 -0.925401 -0.06 0.17 -0.12\\n\",\"xyz\",{\"vibrate\": {\"frames\": 10, \"amplitude\": 1}});\n",
              "\tviewer_16581897837089605.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581897837089605.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581897837089605.animate({\"loop\": \"backAndForth\"});\n",
              "\tviewer_16581897837089605.zoomTo();\n",
              "viewer_16581897837089605.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Animation of the Reaction Coord. at the TS\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/3dmoljs_load.v0": "<div id=\"3dmolviewer_16581897837182486\"  style=\"position: relative; width: 600px; height: 500px\">\n        <p id=\"3dmolwarning_16581897837182486\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n        </div>\n<script>\n\nvar loadScriptAsync = function(uri){\n  return new Promise((resolve, reject) => {\n    var tag = document.createElement('script');\n    tag.src = uri;\n    tag.async = true;\n    tag.onload = () => {\n      resolve();\n    };\n  var firstScriptTag = document.getElementsByTagName('script')[0];\n  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n});\n};\n\nif(typeof $3Dmolpromise === 'undefined') {\n$3Dmolpromise = null;\n  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n}\n\nvar viewer_16581897837182486 = null;\nvar warn = document.getElementById(\"3dmolwarning_16581897837182486\");\nif(warn) {\n    warn.parentNode.removeChild(warn);\n}\n$3Dmolpromise.then(function() {\nviewer_16581897837182486 = $3Dmol.createViewer($(\"#3dmolviewer_16581897837182486\"),{backgroundColor:\"white\"});\nviewer_16581897837182486.zoomTo();\n\tviewer_16581897837182486.addModel(\"4\\n Created by Molecule Class\\nC           0.053355        -0.519118         0.423102\\nO          -0.497209        -1.187230         1.219572\\nC          -0.050587         0.518990        -0.421455\\nO           0.494442         1.187359        -1.221219\\n\");\n\tviewer_16581897837182486.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n\tviewer_16581897837182486.setBackgroundColor(\"0xeeeeee\");\n\tviewer_16581897837182486.zoomTo();\nviewer_16581897837182486.render();\n});\n</script>",
            "text/html": [
              "<div id=\"3dmolviewer_16581897837182486\"  style=\"position: relative; width: 600px; height: 500px\">\n",
              "        <p id=\"3dmolwarning_16581897837182486\" style=\"background-color:#ffcccc;color:black\">You appear to be running in JupyterLab (or JavaScript failed to load for some other reason).  You need to install the 3dmol extension: <br>\n",
              "        <tt>jupyter labextension install jupyterlab_3dmol</tt></p>\n",
              "        </div>\n",
              "<script>\n",
              "\n",
              "var loadScriptAsync = function(uri){\n",
              "  return new Promise((resolve, reject) => {\n",
              "    var tag = document.createElement('script');\n",
              "    tag.src = uri;\n",
              "    tag.async = true;\n",
              "    tag.onload = () => {\n",
              "      resolve();\n",
              "    };\n",
              "  var firstScriptTag = document.getElementsByTagName('script')[0];\n",
              "  firstScriptTag.parentNode.insertBefore(tag, firstScriptTag);\n",
              "});\n",
              "};\n",
              "\n",
              "if(typeof $3Dmolpromise === 'undefined') {\n",
              "$3Dmolpromise = null;\n",
              "  $3Dmolpromise = loadScriptAsync('https://3dmol.org/build/3Dmol.js');\n",
              "}\n",
              "\n",
              "var viewer_16581897837182486 = null;\n",
              "var warn = document.getElementById(\"3dmolwarning_16581897837182486\");\n",
              "if(warn) {\n",
              "    warn.parentNode.removeChild(warn);\n",
              "}\n",
              "$3Dmolpromise.then(function() {\n",
              "viewer_16581897837182486 = $3Dmol.createViewer($(\"#3dmolviewer_16581897837182486\"),{backgroundColor:\"white\"});\n",
              "viewer_16581897837182486.zoomTo();\n",
              "\tviewer_16581897837182486.addModel(\"4\\n Created by Molecule Class\\nC           0.053355        -0.519118         0.423102\\nO          -0.497209        -1.187230         1.219572\\nC          -0.050587         0.518990        -0.421455\\nO           0.494442         1.187359        -1.221219\\n\");\n",
              "\tviewer_16581897837182486.setStyle({\"stick\": {}, \"sphere\": {\"scale\": 0.3}});\n",
              "\tviewer_16581897837182486.setBackgroundColor(\"0xeeeeee\");\n",
              "\tviewer_16581897837182486.zoomTo();\n",
              "viewer_16581897837182486.render();\n",
              "});\n",
              "</script>"
            ]
          },
          "metadata": {}
        },
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Reactant final geometry\n",
            "                Energy          Enthalpy       Gibbs Energy  \n",
            " Delta (D)          -154.49          -152.02          -132.01\n",
            " Eact (D)              0.71            -1.79            -1.79\n",
            " Delta (I)           154.49           152.02           132.01\n",
            " Eact (I)            155.20           150.24           150.24\n",
            "Value of the imaginary frequency at TS:              -100.33\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "hblJc2zeyZPa"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}